{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Basic BigQuery (BBQ): Table Exploration",
      "version": "0.3.2",
      "provenance": [],
      "collapsed_sections": [],
      "toc_visible": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/gist/Gibbsdavidl/dc257e66867a5f3bb8a6c6f351a633c9/isb-cgc-query_of_the_month-november-2018.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "V6hHrkja94oQ",
        "colab_type": "text"
      },
      "source": [
        "# **Basic BigQuery (BBQ): Table Exploration**\n",
        "\n",
        " "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "n97tZmtc93Vf",
        "colab_type": "text"
      },
      "source": [
        "## Configuration and Setup\n",
        "Most of the setup that happens in this section can be kept \"hidden\" when you run this notebook, with the exception of the *authentication* step below which will require that you sign in with Google and then **Allow** the Google Cloud SDK to access your Google account. You will then have to copy/paste a \"verification code\" (which looks something like ``KYuEQglK4/tACQe47cc_GZQa-KGm6JgJY8BiFyX0I27G-0cjnsz7Z6Fka``) into a box below.\n",
        "\n",
        "After completing these steps, if you do not get a \"successfully authenticated\" message, you will not be able to continue running this notebook.  (Because of the nature of Jupyter notebooks, it is also possible that you may see the \"successfully authenticated\" message but it may have been from a previous, expired session, and you may need to re-authenticate)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "wYNMaJPlMEBA",
        "colab_type": "text"
      },
      "source": [
        "### Authenticate with Google\n",
        "Our first step is to authenticate with Google -- you will need to be a member of a Google Cloud Platform (GCP) project, with authorization to run BigQuery jobs in order to run this notebook.  If you don't have access to a GCP project, please contact the ISB-CGC team for help (www.isb-cgc.org)"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "FyrcP8Zs-VkB",
        "colab_type": "code",
        "outputId": "97e392dc-eb8c-486c-f727-047a0f559263",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 34
        }
      },
      "source": [
        "from google.colab import auth\n",
        "try:\n",
        "  auth.authenticate_user()\n",
        "  print('You have been successfully authenticated!')\n",
        "except:\n",
        "  print('You have not been authenticated.')"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "You have been successfully authenticated!\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "jPYmhcmQNA5u",
        "colab_type": "text"
      },
      "source": [
        "### Initialize connection to BigQuery\n",
        "Once you're authenticated, we'll begin getting set up to pull data out of BigQuery.  \n",
        "\n",
        "The first step is to initialize the BigQuery client.  This requires specifying a Google Cloud Platform (GCP) **project id** in which you have the necessary privileges (also referred to as \"roles\") to execute queries and access the data used by this notebook.\n",
        "\n",
        "\n",
        "---\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "92CNdnfsNJUI",
        "colab_type": "code",
        "outputId": "e26c58fb-f839-4eb4-9c2f-5e2d9c845959",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 34
        }
      },
      "source": [
        "from google.cloud import bigquery\n",
        "try:\n",
        "  project_id = 'isb-cgc-bq'\n",
        "  bqclient = bigquery.Client(project=project_id)\n",
        "  print('BigQuery client successfully initialized')\n",
        "except:\n",
        "  print('Failure to initialize BigQuery client')"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "BigQuery client successfully initialized\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "P9UbCekXLt70",
        "colab_type": "text"
      },
      "source": [
        "### Software Configuration"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "m6Z24QXm1r-_",
        "colab_type": "text"
      },
      "source": [
        "#### Libraries & Setup\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Dz-bjj3NQ7-x",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "import logging\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import seaborn as sns\n",
        "import time\n",
        "\n",
        "logging.basicConfig ( level=logging.INFO )"
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "TFCtK69HMEhi",
        "colab_type": "text"
      },
      "source": [
        "#### Python convenience functions\n",
        "\n",
        "We import several 'bbq' convenience functions from github.\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "4rwHtUaMypYS",
        "colab_type": "code",
        "outputId": "81f453e1-a8e8-4d10-bb7f-3794f730efff",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 224
        }
      },
      "source": [
        "!wget https://raw.githubusercontent.com/smrgit/jbox/master/bbq/bbq.py\n",
        "\n",
        "import bbq"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2019-05-30 22:05:03--  https://raw.githubusercontent.com/smrgit/jbox/master/bbq/bbq.py\n",
            "Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.0.133, 151.101.64.133, 151.101.128.133, ...\n",
            "Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.0.133|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 16775 (16K) [text/plain]\n",
            "Saving to: ‘bbq.py’\n",
            "\n",
            "\rbbq.py                0%[                    ]       0  --.-KB/s               \rbbq.py              100%[===================>]  16.38K  --.-KB/s    in 0.01s   \n",
            "\n",
            "2019-05-30 22:05:04 (1.30 MB/s) - ‘bbq.py’ saved [16775/16775]\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xQ3bNGKB10ZG",
        "colab_type": "text"
      },
      "source": [
        "\n",
        "**`bbqRunQuery`**: a relatively generic BigQuery query-execution wrapper function which can be used to run a query in \"dry-run\"  mode or not:  the call to the `query()` function itself is inside a `try/except` block and if it fails we return `None`;  otherwise a \"dry-run\" will return an empty dataframe, and a \"live\" run will return the query results as a dataframe\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "OTQfIJ_S1-XP",
        "colab_type": "text"
      },
      "source": [
        "**`bbqCheckQueryResults`**: a generic function that makes sure that what was returned is a dataframe, and checks how many rows are in the returned dataframe"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "qHvGkaSy279Z",
        "colab_type": "text"
      },
      "source": [
        "**`bbqBuildFieldContentsQuery`**: this python function constructs a query to examine the _contents_ a specific field in the specified table"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "H7_paWIm3OLp",
        "colab_type": "text"
      },
      "source": [
        "**`bbqBuildRepeatedFieldsQuery`**: this python function constructs a query to examine how often a _repeated_ field is repeated in the specified table"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "0JMyKJsw3f2U",
        "colab_type": "text"
      },
      "source": [
        "**`bbqSummarizeQueryResults`**: this python function expects a very specify type of 'query results' and then 'summarizes' them "
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "z-GbZuso3p27",
        "colab_type": "text"
      },
      "source": [
        "**`bbqExploreFieldContents`**: this python function walks through the schema of the specified table in BigQuery and calls previously defined functions to explore the contents of each field"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4f1kJh1w37-y",
        "colab_type": "text"
      },
      "source": [
        "**`bbqExploreRepeatedFields`**: this python function walks through the schema of the specified table in BigQuery and calls previously defined functions to determine how often _repeated_ fields are repeated (even if a field is defined as a _repeated_ field, it may in fact never be repeated more than once -- and it might in fact never be repeated at all!)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "3WS33oVyNEw-",
        "colab_type": "text"
      },
      "source": [
        "## **Exploring a Simple Table**\n",
        "\n",
        "Let's start by exploring a simple table in BigQuery using the functions defined above."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "-pnsSF3PPS-9",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "projectName = \"bigquery-public-data\"\n",
        "datasetName = \"github_repos\"\n",
        "tableName   = \"sample_files\"\n",
        "\n",
        "projectName = \"isb-cgc-bq\"\n",
        "datasetName = \"scratch\"\n",
        "tableName = \"gvar_annot_20190405\""
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "GkcFhcjrQSP8",
        "colab_type": "code",
        "outputId": "3627d789-ec78-4f5c-84ab-2b1454ed55bc",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 479
        }
      },
      "source": [
        "## the \"exploratory\" queries sort and count up unique values, etc, and this type\n",
        "## of operation does not make sense (and can be time-consuming and expensive) \n",
        "## for certain fields (or certain types of fields), so those fields can be \n",
        "## passed in as names or types to be *excluded* using the excludedNames or\n",
        "## excludedTypes lists\n",
        "\n",
        "excludedNames = [ \"path\", \"id\", \"symlink_target\" ]\n",
        "excludedTypes = [ ]\n",
        "xf = bbq.bbqExploreFieldContents ( bqclient, projectName, datasetName, tableName, \n",
        "                              excludedNames, excludedTypes )"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "reference_name                STRING      NULLABLE   (24, 0, 18076178, '10', 3117718, 0.17247661535530354, '1', 'Y', 6, 16)\n",
            "variant_start                 INTEGER     NULLABLE   (329201, 0, 18076178, 97916198, 16470, 0.0009111439376177863, 3935, 249240538, 3879, 26511)\n",
            "variant_end                   INTEGER     NULLABLE   (329201, 0, 18076178, 97916199, 16470, 0.0009111439376177863, 3936, 249240539, 3879, 26511)\n",
            "reference_bases               STRING      NULLABLE   (4, 0, 18076178, 'C', 4676484, 0.2587097781400471, 'A', 'T', 2, 4)\n",
            "variant_quality               FLOAT       NULLABLE   (26898, 5889935, 12186243, 286.8, 70795, 0.005809419687429505, 30.29, 289763.0, 928, 3998)\n",
            "alt                           STRING      NULLABLE   (4, 0, 18076178, 'G', 4568004, 0.2527085095090345, 'A', 'T', 2, 4)\n",
            "aliquot_barcode               STRING      NULLABLE   (10389, 0, 18076178, 'TCGA-W5-AA2J-10A-01D-A46P-09', 32103, 0.0017759838390615538, 'TCGA-02-0003-10A-01D-1490-08', 'TCGA-ZX-AA5X-10A-01D-A42R-09', 1742, 7096)\n",
            "case_barcode                  STRING      NULLABLE   (10389, 0, 18076178, 'TCGA-W5-AA2J', 32103, 0.0017759838390615538, 'TCGA-02-0003', 'TCGA-ZX-AA5X', 1742, 7096)\n",
            "sample_barcode                STRING      NULLABLE   (10389, 0, 18076178, 'TCGA-W5-AA2J-10A', 32103, 0.0017759838390615538, 'TCGA-02-0003-10A', 'TCGA-ZX-AA5X-10A', 1742, 7096)\n",
            "sample_type                   STRING      NULLABLE   (5, 0, 18076178, '10', 14545939, 0.8047021333823997, '01', '14', 1, 2)\n",
            "variant_pValue                FLOAT       NULLABLE   (233322, 0, 18076178, 7.77e-05, 660030, 0.0365138028625299, 0.0, 9.995e-05, 381, 10291)\n",
            "AN                            INTEGER     NULLABLE   (9732, 0, 18076178, 2, 68175, 0.0037715384303031317, 2, 20778, 1286, 3138)\n",
            "BaseQRankSum                  FLOAT       NULLABLE   (12143, 6176488, 11899690, 0.736, 445793, 0.03746257255441108, -21.988, 22.825, 395, 2520)\n",
            "ClippingRankSum               FLOAT       NULLABLE   (5026, 6176488, 11899690, 0.736, 455810, 0.03830435918918896, -4.862, 5.434, 306, 1691)\n",
            "DP                            INTEGER     NULLABLE   (41118, 5889935, 12186243, 5.0, 55463, 0.004551279668393286, 0.0, 17733664.0, 629, 3493)\n",
            "ExcessHet                     FLOAT       NULLABLE   (1, 5889935, 12186243, 3.0103, 12186243, 1.0, 3.0103, 3.0103, 1, 1)\n",
            "FS                            FLOAT       NULLABLE   (14002, 5889935, 12186243, 0.0, 7216093, 0.5921507555692103, 0.0, 671.929, 1, 877)\n",
            "MQ                            FLOAT       NULLABLE   (4357, 5911692, 12164486, 29.0, 842210, 0.06923514894094168, 20.0, 79.04, 469, 2070)\n",
            "MQRankSum                     FLOAT       NULLABLE   (10744, 6176488, 11899690, -0.736, 394901, 0.03318582248781271, -27.487, 17.758, 304, 2363)\n",
            "QD                            FLOAT       NULLABLE   (3843, 5893526, 12182652, 29.8, 101098, 0.008298521537018377, 0.03, 41.12, 562, 2061)\n",
            "ReadPosRankSum                FLOAT       NULLABLE   (8738, 6190033, 11886145, 0.736, 434297, 0.03653808699119858, -24.515, 25.331, 373, 2282)\n",
            "SOR                           FLOAT       NULLABLE   (4507, 5889935, 12186243, 3.611, 683633, 0.056098750041337596, 0.0, 11.427, 52, 849)\n",
            "ADP                           INTEGER     NULLABLE   (1424, 0, 18076178, 5, 1324389, 0.07326709219172327, 4, 7975, 14, 108)\n",
            "TUTE                          FLOAT       NULLABLE   (2534, 0, 18076178, 0.0, 16418988, 0.9083218808754815, 0.0, 0.9723, 1, 1)\n",
            "cytoBand                      STRING      NULLABLE   (831, 0, 18076178, '10q11.21', 2724568, 0.15072699549650373, '10p11.1', 'Yq12', 11, 120)\n",
            "Gene                          STRING      NULLABLE   (24350, 13, 18076165, 'LOC441666', 2723854, 0.1506876043674087, 'A1BG', 'ZZZ3', 13, 542)\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ZLg9hzSucIAV",
        "colab_type": "code",
        "outputId": "e339faac-6216-4efe-959e-faac76dd7bb3",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 34
        }
      },
      "source": [
        "xr = bbq.bbqExploreRepeatedFields ( bqclient, projectName, datasetName, tableName )"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            " no REPEATED fields found in this table\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GcELzh4qdDLv",
        "colab_type": "text"
      },
      "source": [
        "## **Exploring a Complex Table**\n",
        "\n",
        "Now for a table with a significantly more complicated schema which includes RECORDs as well as REPEATED fields."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "7-Mc2VRLdCZO",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "projectName = \"bigquery-public-data\"\n",
        "datasetName = \"github_repos\"\n",
        "tableName = \"sample_commits\"\n",
        "\n",
        "projectName = \"isb-cgc-secure-testing\"\n",
        "datasetName = \"VT_test\"\n",
        "tableName = \"SixteenVCFs_24mayH\""
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "BiY8P6RFdUMv",
        "colab_type": "code",
        "outputId": "e3f74d3b-d69c-4e1b-c685-210e35dfc225",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 2026
        }
      },
      "source": [
        "excludedNames = [ ]\n",
        "excludedTypes = [ ]\n",
        "xf = bbq.bbqExploreFieldContents ( bqclient, \n",
        "                                   projectName, datasetName, tableName, \n",
        "                                   excludedNames, excludedTypes )"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "reference_name                STRING      NULLABLE   (25, 0, 316773, 'chr1', 31306, 0.09882786727404166, 'chr1', 'chrY', 9, 18)\n",
            "start_position                INTEGER     NULLABLE   (178278, 0, 316773, 24749170, 8, 2.5254677639824102e-05, 267, 248918270, 43946, 146600)\n",
            "end_position                  INTEGER     NULLABLE   (178054, 0, 316773, 24749171, 8, 2.5254677639824102e-05, 270, 248918271, 43943, 146376)\n",
            "reference_bases               STRING      NULLABLE   (4508, 0, 316773, 'G', 103728, 0.3274521502779593, 'A', 'TTTTTTTTGAG', 2, 4)\n",
            "alternate_bases               RECORD      REPEATED   [2] \n",
            "    > alt                     STRING      NULLABLE   (12243, 0, 317065, 'A', 103930, 0.3277876776055383, 'A', 'TTTTTTTTTTG', 2, 4)\n",
            "    > CSQ                     RECORD      REPEATED   [66] \n",
            "        > allele              STRING      NULLABLE   (10639, 0, 2023917, 'A', 657008, 0.32462200772067235, '-', 'TTTTTTTTTTTT', 2, 5)\n",
            "        > Consequence         STRING      NULLABLE   (119, 0, 2023917, 'intron_variant', 630892, 0.31171831651199133, '3_prime_UTR_variant', 'upstream_gene_variant', 3, 9)\n",
            "        > IMPACT              STRING      NULLABLE   (4, 0, 2023917, 'MODIFIER', 1715744, 0.8477343685536511, 'HIGH', 'MODIFIER', 1, 2)\n",
            "        > SYMBOL              STRING      NULLABLE   (34960, 83152, 1940765, 'TTN-AS1', 13733, 0.007076075671191515, '5S_rRNA', 'uc_338', 2996, 12493)\n",
            "        > Gene                STRING      NULLABLE   (35547, 83152, 1940765, 'ENSG00000237298', 13733, 0.007076075671191515, 'ENSG00000000003', 'ENSG00000281910', 3007, 12586)\n",
            "        > Feature_type        STRING      NULLABLE   (3, 10524, 2013393, 'Transcript', 1940765, 0.9639275591004837, 'MotifFeature', 'Transcript', 1, 1)\n",
            "        > Feature             STRING      NULLABLE   (177764, 10524, 2013393, 'ENST00000342175', 876, 0.00043508644363023015, 'ENSR00000000105', 'MA0541.1', 31067, 107964)\n",
            "        > BIOTYPE             STRING      NULLABLE   (50, 10738, 2013179, 'protein_coding', 1138223, 0.565385889679954, '3prime_overlapping_ncrna', 'unprocessed_pseudogene', 1, 5)\n",
            "        > EXON                STRING      NULLABLE   (4036, 1578216, 445701, '1/1', 21015, 0.047150443907462626, '1/1', '99/363', 38, 409)\n",
            "        > INTRON              STRING      NULLABLE   (5276, 1064368, 959549, '2/2', 27353, 0.028506100261685437, '1/1', '99/312', 43, 431)\n",
            "        > HGVSc               STRING      NULLABLE   (783432, 618932, 1404985, 'ENST00000259241.6:c.754C>T', 6, 4.27050822606647e-06, 'ENST00000000233.8:c.259-230C>A', 'ENST00000631366.1:n.556-8700T>C', 194044, 642933)\n",
            "        > HGVSp               STRING      NULLABLE   (125685, 1744207, 279710, 'ENSP00000259241.6:p.Arg252Cys', 6, 2.145078831647063e-05, 'ENSP00000000442.6:p.Gly88Ala', 'ENST00000631357.1:c.1140G>A(p.%3D)', 35597, 97714)\n",
            "        > cDNA_position       STRING      NULLABLE   (210689, 83152, 1940765, '-/582', 5148, 0.002652562262819043, '-/100', '?-963/1988', 1992, 74386)\n",
            "        > CDS_position        STRING      NULLABLE   (109393, 698231, 1325686, '-/444', 2535, 0.001912217523606646, '-/10', '?-967/1721', 705, 33295)\n",
            "        > Protein_position    STRING      NULLABLE   (99572, 698231, 1325686, '-/147', 3021, 0.002278820173102831, '-/1', '?-9713/35991', 535, 29926)\n",
            "        > Amino_acids         STRING      NULLABLE   (3598, 1744187, 279730, 'R/Q', 10490, 0.03750044685947163, '*', 'YY/*', 23, 113)\n",
            "        > Codons              STRING      NULLABLE   (5107, 1744187, 279730, 'cGa/cAa', 7052, 0.02521002395166768, '-/A', 'ttt/ttTTCt', 47, 338)\n",
            "        > Existing_variation  STRING      NULLABLE   (39742, 1541551, 482366, 'rs758327185', 300, 0.0006219343817764933, '2010-08-Y-3546&rs796510539', 'var_9_134049453', 5926, 22932)\n",
            "        > ALLELE_NUM          STRING      NULLABLE   (3, 0, 2023917, '1', 2022676, 0.9993868325627978, '1', '3', 1, 1)\n",
            "        > DISTANCE            STRING      NULLABLE   (5001, 1488173, 535744, '42', 442, 0.0008250209055071079, '0', '999', 1936, 4248)\n",
            "        > STRAND              STRING      NULLABLE   (2, 82938, 1940979, '1', 1004422, 0.5174821572000521, '-1', '1', 1, 2)\n",
            "        > FLAGS               STRING      NULLABLE   (4, 1734545, 289372, 'cds_end_NF', 154398, 0.5335623349874902, 'cds_end_NF', 'cds_start_NF&cds_end_NF', 1, 3)\n",
            "        > VARIANT_CLASS       STRING      NULLABLE   (3, 0, 2023917, 'SNV', 1771345, 0.8752063449242237, 'SNV', 'insertion', 1, 2)\n",
            "        > SYMBOL_SOURCE       STRING      NULLABLE   (7, 83152, 1940765, 'HGNC', 1843102, 0.9496780908559254, 'Clone_based_ensembl_gene', 'miRBase', 1, 1)\n",
            "        > HGNC_ID             STRING      NULLABLE   (25295, 179489, 1844428, 'HGNC:44124', 13733, 0.007445668792709718, 'HGNC:100', 'HGNC:9999', 2758, 10622)\n",
            "        > CANONICAL           STRING      NULLABLE   (1, 1595006, 428911, 'YES', 428911, 1.0, 'YES', 'YES', 1, 1)\n",
            "        > TSL                 STRING      NULLABLE   (5, 204462, 1819455, '1', 636665, 0.3499207180172085, '1', '5', 2, 4)\n",
            "        > ENSP                STRING      NULLABLE   (87354, 698231, 1325686, 'ENSP00000340554', 876, 0.0006607899608202847, 'ENSP00000000233', 'ENSP00000487592', 17019, 56566)\n",
            "        > SWISSPROT           STRING      NULLABLE   (17619, 1334328, 689589, 'Q8WZ42', 4530, 0.006569130308053057, 'A0A087WVF3', 'Q9Y6Z7', 2322, 9539)\n",
            "        > TREMBL              STRING      NULLABLE   (47880, 1331811, 692106, 'W4VSQ6', 1166, 0.0016847130352865025, 'A0A023T6R1', 'X6RM59', 9039, 31032)\n",
            "        > UNIPARC             STRING      NULLABLE   (69188, 833301, 1190616, 'UPI00025287CD', 1752, 0.001471507186196053, 'UPI000000003C', 'UPI0005C1235D', 12437, 43926)\n",
            "        > RefSeq              STRING      NULLABLE   (32772, 1434930, 588987, 'NM_001256850.1', 876, 0.001487299380122142, '28432', 'NM_214711.3', 6436, 21087)\n",
            "        > SIFT                STRING      NULLABLE   (203, 1857859, 166058, 'deleterious(0)', 48751, 0.29357814739428395, 'deleterious(0)', 'tolerated_low_confidence(1)', 5, 57)\n",
            "        > PolyPhen            STRING      NULLABLE   (1002, 1848072, 175845, 'probably_damaging(1)', 10277, 0.058443515596121584, 'benign(0)', 'unknown(0)', 58, 607)\n",
            "        > DOMAINS             STRING      NULLABLE   (10719, 1827248, 196669, 'Low_complexity_(Seg):Seg', 7476, 0.03801310831905384, 'Cleavage_site_(Signalp):Sigp', 'Transmembrane_helices:Tmhmm&TIGRFAM_domain:TIGR01657&Superfamily_domains:SSF81665', 493, 4704)\n",
            "        > HGVS_OFFSET         STRING      NULLABLE   (45, 1996411, 27506, '1', 12655, 0.46008143677743035, '1', '9', 2, 12)\n",
            "        > GMAF                STRING      NULLABLE   (5992, 1882301, 141616, 'A:0.0002', 21675, 0.1530547395774489, '-:0.0002', 'TTTTG:0.1993', 17, 2129)\n",
            "        > AFR_MAF             STRING      NULLABLE   (4154, 1866895, 157022, 'A:0.0000', 26596, 0.16937753945307027, '-:0', 'TTTTTTC:0.3616', 8, 1353)\n",
            "        > AMR_MAF             STRING      NULLABLE   (2709, 1866946, 156971, 'A:0.0000', 34672, 0.2208815641105682, '-:0', 'TTTTTTC:0.1902', 4, 897)\n",
            "        > EAS_MAF             STRING      NULLABLE   (3502, 1867086, 156831, 'A:0.0000', 35020, 0.22329768986998744, '-:0', 'TTTTTTC:0.2341', 4, 1053)\n",
            "        > EUR_MAF             STRING      NULLABLE   (3470, 1866969, 156948, 'T:0.0000', 34533, 0.22002828962458904, '-:0', 'TTTTTTC:0.2485', 4, 1133)\n",
            "        > SAS_MAF             STRING      NULLABLE   (3393, 1866996, 156921, 'A:0.0000', 33295, 0.21217682783056444, '-:0', 'TTTTTTC:0.4192', 4, 1067)\n",
            "        > AA_MAF              STRING      NULLABLE   (970, 1940413, 83504, 'T:0', 16724, 0.2002778310021077, '-:0.0005', 'T:0.8756', 4, 157)\n",
            "        > EA_MAF              STRING      NULLABLE   (821, 1940413, 83504, 'T:0.0001', 14488, 0.1735006706265568, '-:0.0001', 'T:0.7214', 4, 70)\n",
            "        > ExAC_MAF            STRING      NULLABLE   (9083, 1696099, 327818, 'A:8.236e-06', 12903, 0.039360254775515684, '-:0.010', 'TTTTGG:8.237e-06', 199, 3754)\n",
            "        > ExAC_Adj_MAF        STRING      NULLABLE   (15124, 1696116, 327801, 'A:1.648e-05', 2415, 0.007367274657490368, '-:0', 'TTTTGG:8.33e-06', 1452, 8227)\n",
            "        > ExAC_AFR_MAF        STRING      NULLABLE   (10311, 1696173, 327744, 'A:0', 71703, 0.218777460456942, '-:0', 'TTTTGG:0', 7, 4228)\n",
            "        > ExAC_AMR_MAF        STRING      NULLABLE   (7446, 1696188, 327729, 'A:0', 76490, 0.23339405423383344, '-:0', 'TTTTGG:0', 4, 2437)\n",
            "        > ExAC_EAS_MAF        STRING      NULLABLE   (7134, 1696169, 327748, 'A:0', 80225, 0.2447764746085407, '-:0', 'TTTTGG:0', 3, 2164)\n",
            "        > ExAC_FIN_MAF        STRING      NULLABLE   (7909, 1696333, 327584, 'A:0', 94639, 0.2888999462733223, '-:0', 'TTTTGG:0', 2, 2356)\n",
            "        > ExAC_NFE_MAF        STRING      NULLABLE   (12057, 1696120, 327797, 'A:0', 35193, 0.10736217842140106, '-:0', 'TTTTGG:1.513e-05', 315, 5615)\n",
            "        > ExAC_OTH_MAF        STRING      NULLABLE   (5510, 1696218, 327699, 'A:0', 95873, 0.2925642128904879, '-:0', 'TTTTGG:0', 2, 1258)\n",
            "        > ExAC_SAS_MAF        STRING      NULLABLE   (6872, 1696198, 327719, 'T:0', 66895, 0.20412304443746015, '-:0', 'TTTTGG:0', 4, 2163)\n",
            "        > CLIN_SIG            STRING      NULLABLE   (20, 2018432, 5485, 'pathogenic', 1544, 0.28149498632634457, 'benign', 'uncertain_significance&pathogenic', 3, 7)\n",
            "        > SOMATIC             STRING      NULLABLE   (2, 2023598, 319, '1&0', 176, 0.5517241379310345, '1', '1&0', 1, 2)\n",
            "        > PHENO               STRING      NULLABLE   (2, 2017773, 6144, '1', 5822, 0.9475911458333334, '1', '1&0', 1, 1)\n",
            "        > PUBMED              STRING      NULLABLE   (351, 2020113, 3804, '21499247', 187, 0.049158780231335436, '10712197', '26487440&23679094', 57, 218)\n",
            "        > MOTIF_NAME          STRING      NULLABLE   (24, 2023703, 214, 'Name/Accession_association_Znf263:MA0528.1', 38, 0.17757009345794392, 'Accession_association_Jund:MA0492.1', 'Name/Accession_association_Znf263:MA0528.1', 5, 14)\n",
            "        > MOTIF_POS           STRING      NULLABLE   (23, 2023703, 214, '5', 33, 0.1542056074766355, '-1', '9', 5, 12)\n",
            "        > HIGH_INF_POS        STRING      NULLABLE   (2, 2023703, 214, 'N', 146, 0.6822429906542056, 'N', 'Y', 1, 2)\n",
            "        > MOTIF_SCORE_CHANGE  STRING      NULLABLE   (61, 2023729, 188, '-0.214', 9, 0.047872340425531915, '-0.003', '0.058', 15, 42)\n",
            "        > ENTREZ              STRING      NULLABLE   (25837, 169115, 1854802, '100506866', 13733, 0.007404024796177705, '1', '9997', 2779, 10758)\n",
            "        > EVIDENCE            STRING      NULLABLE   (60, 1571544, 452373, 'E_ExAC', 170025, 0.37585134391309827, 'E_Cited', 'E_Phenotype_or_Disease&E_ExAC', 3, 8)\n",
            "names                         STRING      REPEATED   (34298, 0, 45775, 'rs201299133', 4, 8.738394320043692e-05, 'rs1000042', 'rs9997310', 11410, 29720)\n",
            "filter                        STRING      REPEATED   (30, 0, 489019, 'PASS', 143924, 0.2943116729615823, 'DETP20', 'triallelic_site', 3, 10)\n",
            "call                          RECORD      REPEATED   [28] \n",
            "    > name                    STRING      NULLABLE   (32, 0, 633546, 'TCGA-AP-A0LM-01A-11D-A122-09;tumor;mutect2', 138016, 0.2178468493211227, 'TCGA-44-4112-01A-01D-1458-08;tumor;muse', 'TCGA-AP-A0LM-10A-01D-A122-09;normal;varscan2', 3, 10)\n",
            "    > genotype                INTEGER     REPEATED   (4, 0, 1267127, 0, 947122, 0.747456253398436, 0, 3, 1, 2)\n",
            "    > AMQ                     INTEGER     REPEATED   (69, 0, 117137, 60, 91759, 0.7833477039705644, 0, 70, 1, 8)\n",
            "    > BCOUNT                  INTEGER     REPEATED   (982, 0, 321824, 0, 178299, 0.5540264243810281, 0, 9707, 1, 74)\n",
            "    > BQ                      INTEGER     REPEATED   (40, 0, 345191, 0, 49503, 0.14340756276959712, 0, 39, 5, 13)\n",
            "    > DP                      INTEGER     NULLABLE   (1333, 332260, 301286, 13.0, 3055, 0.010139867103018394, 1.0, 10068.0, 72, 280)\n",
            "    > DP4                     STRING      REPEATED   (795, 0, 749424, '0', 236432, 0.3154849591152672, '0', '996', 7, 80)\n",
            "    > GQ                      INTEGER     NULLABLE   (256, 553090, 80456, 255.0, 16096, 0.2000596599383514, 0.0, 255.0, 29, 142)\n",
            "    > IGT                     STRING      NULLABLE   (7, 553090, 80456, '0/0', 40083, 0.49819777269563487, '0/0', '2/2', 2, 2)\n",
            "    > JGQ                     INTEGER     NULLABLE   (201, 553090, 80456, 29.0, 1554, 0.01931490504126479, 25.0, 225.0, 43, 126)\n",
            "    > MQ                      INTEGER     NULLABLE   (67, 553090, 80456, 60.0, 62526, 0.7771452719498857, 3.0, 70.0, 1, 8)\n",
            "    > SS                      INTEGER     NULLABLE   (3, 496125, 137421, 2.0, 97193, 0.7072645374433311, 0.0, 2.0, 1, 2)\n",
            "    > SSC                     INTEGER     NULLABLE   (201, 593318, 40228, 25.0, 2805, 0.06972755294819528, 25.0, 225.0, 30, 119)\n",
            "    > VAQ                     INTEGER     NULLABLE   (244, 553090, 80456, 0.0, 40081, 0.49817291438798844, 0.0, 255.0, 2, 101)\n",
            "    > AD                      INTEGER     REPEATED   (1607, 0, 999472, 0, 236726, 0.23685105735828518, 0, 8669, 11, 142)\n",
            "    > AF                      FLOAT       NULLABLE   (1316, 301286, 332260, 0.0, 146748, 0.4416661650514657, 0.0, 1.0, 7, 183)\n",
            "    > ALT_F1R2                INTEGER     NULLABLE   (181, 301286, 332260, 0.0, 154779, 0.46583699512430027, 0.0, 552.0, 2, 10)\n",
            "    > ALT_F2R1                INTEGER     NULLABLE   (182, 301286, 332260, 0.0, 212467, 0.6394600613977006, 0.0, 532.0, 1, 9)\n",
            "    > FOXOG                   FLOAT       NULLABLE   (610, 483591, 149955, 0.0, 24982, 0.16659664566036478, 0.0, 1.0, 5, 98)\n",
            "    > PGT                     STRING      NULLABLE   (2, 520760, 112786, '0|1', 112236, 0.9951235082368379, '0|1', '1|0', 1, 1)\n",
            "    > PID                     STRING      NULLABLE   (21943, 520760, 112786, '143961020_AAG_A', 34, 0.0003014558544500204, '100002338_TA_T', '99983471_GCA_G', 7964, 19124)\n",
            "    > QSS                     INTEGER     REPEATED   (19510, 0, 664520, 0, 142459, 0.21437879973514717, 0, 246307, 253, 4250)\n",
            "    > REF_F1R2                INTEGER     NULLABLE   (965, 301286, 332260, 2.0, 12753, 0.03838259194606634, 0.0, 4307.0, 22, 114)\n",
            "    > REF_F2R1                INTEGER     NULLABLE   (962, 301286, 332260, 2.0, 11664, 0.03510503822307831, 0.0, 4362.0, 23, 114)\n",
            "    > FREQ                    STRING      NULLABLE   (3097, 526646, 106900, '0%', 45722, 0.4277081384471469, '0%', '92.94%', 10, 848)\n",
            "    > RD                      INTEGER     NULLABLE   (818, 526646, 106900, 18.0, 1271, 0.011889616463985033, 1.0, 2117.0, 57, 231)\n",
            "ACGTNacgtnMINUS               INTEGER     REPEATED   (742, 0, 2754150, 0, 1861926, 0.6760437884646806, 0, 2268, 1, 15)\n",
            "ACGTNacgtnPLUS                INTEGER     REPEATED   (741, 0, 2754150, 0, 1859402, 0.6751273532668881, 0, 2186, 1, 15)\n",
            "SOMATIC                       BOOLEAN     NULLABLE   (1, 206358, 110415, True, 110415, 1.0, True, True, 1, 1)\n",
            "DB                            BOOLEAN     NULLABLE   (1, 284280, 32493, True, 32493, 1.0, True, True, 1, 1)\n",
            "ECNT                          STRING      NULLABLE   (22, 150643, 166130, '1', 88564, 0.5331005838800938, '1', '9', 1, 5)\n",
            "HCNT                          STRING      NULLABLE   (100, 150643, 166130, '1', 47708, 0.2871726960813821, '1', '96', 3, 19)\n",
            "MAX_ED                        INTEGER     NULLABLE   (243, 239207, 77566, 2.0, 5417, 0.06983729984787149, 1.0, 292.0, 17, 109)\n",
            "MIN_ED                        INTEGER     NULLABLE   (231, 239207, 77566, 1.0, 15102, 0.1946987081968904, 1.0, 247.0, 4, 68)\n",
            "NLOD                          STRING      NULLABLE   (15014, 150643, 166130, '1.20', 4800, 0.02889303557455005, '0.501', '99.99', 636, 6696)\n",
            "RPA                           INTEGER     REPEATED   (29, 0, 11188, 1, 2103, 0.18796925277082588, -1, 32, 4, 9)\n",
            "RU                            STRING      NULLABLE   (308, 311179, 5594, 'A', 1314, 0.23489452985341439, 'A', 'TTTTTTG', 3, 20)\n",
            "STR                           BOOLEAN     NULLABLE   (1, 311179, 5594, True, 5594, 1.0, True, True, 1, 1)\n",
            "TLOD                          STRING      NULLABLE   (11947, 150643, 166130, '6.83', 855, 0.005146571961716728, '10.00', '99.99', 375, 3381)\n",
            "DP                            INTEGER     NULLABLE   (1233, 263323, 53450, 72.0, 287, 0.005369504209541628, 16.0, 3684.0, 137, 433)\n",
            "GPV                           FLOAT       NULLABLE   (1, 263323, 53450, 1.0, 53450, 1.0, 1.0, 1.0, 1, 1)\n",
            "SPV                           FLOAT       NULLABLE   (43994, 263323, 53450, 0.0, 109, 0.002039289055191768, 0.0, 0.069968, 17269, 38649)\n",
            "SS                            STRING      NULLABLE   (1, 263323, 53450, '2', 53450, 1.0, '2', '2', 1, 1)\n",
            "SSC                           STRING      NULLABLE   (245, 263323, 53450, '13', 1755, 0.03283442469597755, '100', '99', 32, 122)\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "EL3wTf8y9hNd",
        "colab_type": "code",
        "outputId": "ab5b2576-5ee0-4e86-f222-d4686e19ce00",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 309
        }
      },
      "source": [
        "xr = bbq.bbqExploreRepeatedFields ( bqclient, projectName, datasetName, tableName )"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "alternate_bases               RECORD      REPEATED   (3, 0, 316773, 1, 316484, 0.9990876747702614, 1, 3, 1, 1)\n",
            "    > CSQ                     RECORD      REPEATED   (72, 0, 317065, 3, 38056, 0.12002586220491067, 1, 125, 5, 13)\n",
            "names                         STRING      REPEATED   (2, 0, 316773, 0, 270998, 0.8554958913796314, 0, 1, 1, 2)\n",
            "filter                        STRING      REPEATED   (7, 0, 316773, 1, 214380, 0.6767622240531863, 1, 7, 1, 3)\n",
            "call                          RECORD      REPEATED   always repeated 2 time(s)\n",
            "    > genotype                INTEGER     REPEATED   (2, 0, 633546, 2, 633511, 0.9999447553926629, 2, 3, 1, 1)\n",
            "    > AMQ                     INTEGER     REPEATED   (3, 0, 633546, 0, 553090, 0.8730068534881446, 0, 2, 1, 2)\n",
            "    > BCOUNT                  INTEGER     REPEATED   (2, 0, 633546, 0, 553090, 0.8730068534881446, 0, 4, 1, 2)\n",
            "    > BQ                      INTEGER     REPEATED   (4, 0, 633546, 0, 439160, 0.693177764519072, 0, 3, 1, 2)\n",
            "    > DP4                     STRING      REPEATED   (2, 0, 633546, 0, 446190, 0.7042740385070697, 0, 4, 1, 2)\n",
            "    > AD                      INTEGER     REPEATED   (4, 0, 633546, 2, 445995, 0.7039662471233343, 0, 3, 1, 3)\n",
            "    > PL                      INTEGER     REPEATED   always repeated 0 time(s)\n",
            "    > QSS                     INTEGER     REPEATED   (2, 0, 633546, 2, 332260, 0.5244449495379973, 0, 2, 1, 2)\n",
            "ACGTNacgtnMINUS               INTEGER     REPEATED   (2, 0, 316773, 10, 275415, 0.8694396302715194, 0, 10, 1, 2)\n",
            "ACGTNacgtnPLUS                INTEGER     REPEATED   (2, 0, 316773, 10, 275415, 0.8694396302715194, 0, 10, 1, 2)\n",
            "RPA                           INTEGER     REPEATED   (2, 0, 316773, 0, 311179, 0.982340666660353, 0, 2, 1, 1)\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "N9GQYflzapa9",
        "colab_type": "text"
      },
      "source": [
        "## **Platinum Genomes**\n",
        "\n",
        "The \"*Platinum Genomes*\" are a set of sequences derived from the three-generation, 17-member CEPH 1463 pedigree.  This reference data set is open-access and is described in detail in the Genome Research [publication](https://genome.cshlp.org/content/27/1/157) by Eberle *et al*.\n",
        "\n",
        "In this section, we're going use the previously defined functions to have a look at the Platinum Genome variants *called by Google's DeepVariant method* which have been made available in a publicly accessible [table](https://bigquery.cloud.google.com/table/bigquery-public-data:human_genome_variants.platinum_genomes_deepvariant_variants_20180823?pli=1&tab=schema) in BigQuery.  \n",
        "\n",
        "The size of this table is **19.6 GB** and it has nearly 106 *million* rows of data.  It was created in August 2018.  More information can be found by clicking on the `Details` button in the BigQuery web UI."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "LIu759ujNBhx",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "projectName = \"bigquery-public-data\"\n",
        "datasetName = \"human_genome_variants\"\n",
        "tableName = \"platinum_genomes_deepvariant_variants_20180823\""
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "lYfZmda6X1ZE",
        "colab_type": "text"
      },
      "source": [
        "The **exploreFieldContents** function will walk through the entire table schema and output one row per field, with the following information:\n",
        "- field name\n",
        "- field type (*eg* STRING, INTEGER, RECORD)\n",
        "- field mode (*eg* NULLABLE, REPEATED)\n",
        "- finally comes a summary of the *contents* of this field:\n",
        "  - if this field is a RECORD, the next field will be a number in brackets, indicating the number of *child* fields within this RECORD\n",
        "  - otherwise, the following 10 values describe the contents of this field:\n",
        "    - total number of _unique_ non-null values observed\n",
        "    - total number of null values observed\n",
        "    - total number of non-null values observed\n",
        "    - single most common (non-null) value\n",
        "    - number of times the most common value is observed\n",
        "    - fraction of total non-null values that are equal to this most common value\n",
        "    - 'minimum' observed value (based on numeric or alphabetic sorting)\n",
        "    - 'maximum' observed value (based on numeric or alphabetic sorting)\n",
        "    - number of distinct values that represent at least 50% of non-null values\n",
        "    - number of distinct values that represent at least 90% of non-null values\n",
        "    \n",
        "**NB**: fields which are always *null* will not be shown at all in this output.  For example, for this table, the **call** record contains 12 fields: name, genotype, phaseset, GQ, DP, MIN_DP, AD, VAF, GL, PL, quality, and filter.  But two of these fields (phaseset, and GL) are always null (or missing), and are therefore not contained in the output generated by **exploreFieldContents**."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "VSvXbvG2qpDg",
        "colab_type": "code",
        "outputId": "1631e29b-e353-43e7-c952-15cc13bc9590",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 377
        }
      },
      "source": [
        "## the \"exploratory\" queries sort and count up unique values, etc, and this\n",
        "## type of operation does not make sense (and can be very time-consuming) for\n",
        "## certain fields, so those fields can be passed in as names to be *excluded*\n",
        "## using the `excludedNames` list:\n",
        "\n",
        "excludedNames = [ \"start_position\", \"end_position\" ]\n",
        "excludedTypes = [ ]\n",
        "xf = bbq.bbqExploreFieldContents ( bqclient, \n",
        "                                   projectName, datasetName, tableName, \n",
        "                                   excludedNames, excludedTypes )"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "reference_name                STRING      NULLABLE   (24, 0, 105923159, 'chrX', 14598875, 0.13782514737877105, 'chr1', 'chrY', 8, 19)\n",
            "start_position                INTEGER     NULLABLE   (this field was excluded)\n",
            "end_position                  INTEGER     NULLABLE   (this field was excluded)\n",
            "reference_bases               STRING      NULLABLE   (67253, 0, 105923159, 'T', 28081226, 0.26510940822676937, 'A', 'TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTGTTGGAG', 2, 4)\n",
            "alternate_bases               RECORD      REPEATED   [1] \n",
            "    > alt                     STRING      NULLABLE   (52211, 0, 118997285, '<*>', 105923159, 0.8901308882803503, '<*>', 'TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT', 1, 2)\n",
            "quality                       FLOAT       NULLABLE   (830, 0, 105923159, 0.0, 97211191, 0.9177519998247031, 0.0, 150.0, 1, 1)\n",
            "filter                        STRING      REPEATED   (2, 0, 13234436, 'PASS', 8117976, 0.6133979566639636, 'PASS', 'RefCall', 1, 2)\n",
            "call                          RECORD      REPEATED   [12] \n",
            "    > name                    STRING      NULLABLE   (6, 0, 182104652, 'NA12891', 33487348, 0.1838906784215485, 'NA12877', 'NA12892', 3, 6)\n",
            "    > genotype                INTEGER     REPEATED   (5, 0, 364209304, 0, 326217787, 0.8956876812790043, 0, 4, 1, 2)\n",
            "    > GQ                      INTEGER     NULLABLE   (70, 0, 182104652, 50, 74660263, 0.409985479118897, 0, 150, 3, 27)\n",
            "    > DP                      INTEGER     NULLABLE   (1248, 143555264, 38549388, 50.0, 1190633, 0.030885911859352994, 2.0, 1426.0, 19, 56)\n",
            "    > MIN_DP                  INTEGER     NULLABLE   (1271, 38549388, 143555264, 16.0, 5869036, 0.04088346074164163, 0.0, 1426.0, 21, 48)\n",
            "    > AD                      INTEGER     REPEATED   (1123, 0, 115998659, 0, 48031360, 0.4140682350474414, 0, 1309, 6, 42)\n",
            "    > VAF                     FLOAT       REPEATED   (35070, 0, 77449271, 0.0, 38549388, 0.49773726081940783, 0.0, 1.0, 2, 406)\n",
            "    > PL                      INTEGER     REPEATED   (920, 0, 663367483, 0, 183879094, 0.277190394030815, 0, 2999, 7, 121)\n",
            "    > quality                 FLOAT       NULLABLE   (830, 0, 182104652, 0.0, 153701513, 0.8440284820400964, 0.0, 150.0, 1, 115)\n",
            "    > filter                  STRING      REPEATED   (2, 0, 38549388, 'PASS', 26867854, 0.6969722580291028, 'PASS', 'RefCall', 1, 2)\n",
            "partition_date_please_ignore  DATE        NULLABLE   (1, 0, 105923159, datetime.date(2018, 8, 23), 105923159, 1.0, datetime.date(2018, 8, 23), datetime.date(2018, 8, 23), 1, 1)\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "4r7B7vTirf6q",
        "colab_type": "code",
        "outputId": "488ed2fd-58af-4b46-947e-a5c2c47ea852",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 207
        }
      },
      "source": [
        "xr = bbq.bbqExploreRepeatedFields ( bqclient, \n",
        "                                    projectName, datasetName, tableName )"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "alternate_bases               RECORD      REPEATED   (5, 0, 105923159, 1, 93106707, 0.8790023624578643, 1, 5, 1, 2)\n",
            "names                         STRING      REPEATED   always repeated 0 time(s)\n",
            "filter                        STRING      REPEATED   (3, 0, 105923159, 0, 93106707, 0.8790023624578643, 0, 2, 1, 2)\n",
            "call                          RECORD      REPEATED   (6, 0, 105923159, 1, 72441909, 0.683910012540317, 1, 6, 1, 4)\n",
            "    > genotype                INTEGER     REPEATED   always repeated 2 time(s)\n",
            "    > AD                      INTEGER     REPEATED   (5, 0, 182104652, 0, 143555264, 0.7883118988086038, 0, 6, 1, 2)\n",
            "    > VAF                     FLOAT       REPEATED   (5, 0, 182104652, 0, 143555264, 0.7883118988086038, 0, 5, 1, 2)\n",
            "    > GL                      FLOAT       REPEATED   always repeated 0 time(s)\n",
            "    > PL                      INTEGER     REPEATED   (5, 0, 182104652, 3, 143555264, 0.7883118988086038, 3, 21, 1, 2)\n",
            "    > filter                  STRING      REPEATED   (2, 0, 182104652, 0, 143555264, 0.7883118988086038, 0, 1, 1, 2)\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "uZn8-D6zoUbn",
        "colab_type": "text"
      },
      "source": [
        "The variants in this table are not annotated, so the contents of the table are  relatively simple."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "RO4dUMTAoa9g",
        "colab_type": "text"
      },
      "source": [
        "## *gnomAD Exome data: hg19 release 2.0.2*\n",
        "\n",
        "In this section, we'll explore data released by the [gnomAD project](https://gnomad.broadinstitute.org/).  Public data files are available from a Google Cloud Storage bucket ([gs://gnomad-public](https://console.cloud.google.com/storage/browser/gnomad-public/))\n",
        "\n",
        "The r2.0.2 exome VCF file is available in GCS at `gs://gnomad-public/release/2.0.2/vcf/exomes/gnomad.exomes.r2.0.2.sites.vcf.bgz`\n",
        "\n",
        "The key difference between tables in Google's **human_variant_annotation** and **human_genome_variants** is whether or not the variants have been annotated using the [**Variant Effect Predictor**](https://uswest.ensembl.org/info/docs/tools/vep/vep_formats.html) option in Google's [**Variant Transforms**](https://cloud.google.com/genomics/docs/how-tos/variant-transforms) tool.\n",
        "\n",
        "These annotations *dramatically* increase the size of the destination table in BigQuery.  The size of this gnomAD table is **68 GB** and contains ~15 million rows, as compared to the Platinum Genomes table which was less than 20 GB, with nearly 106 million rows.\n",
        "\n",
        "\n",
        "(Note that the most recent version of gnomAD is release 2.1)"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "6utSUQ5mqjcp",
        "colab_type": "code",
        "colab": {}
      },
      "source": [
        "projectName = \"bigquery-public-data\"\n",
        "datasetName = \"human_variant_annotation\"\n",
        "tableName = \"gnomad_gnomad_exomes_hg19_release_2_0_2\""
      ],
      "execution_count": 0,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ECYgh_f3-tP7",
        "colab_type": "text"
      },
      "source": [
        "We can use the `bq` command-line tool to show information about this table:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "sY8QTUHqOdYq",
        "colab_type": "code",
        "outputId": "2e3915f0-dbbd-490c-96fe-238d09527a52",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 119
        }
      },
      "source": [
        "!gcloud config set project 'isb-cgc-bq'"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Updated property [core/project].\n",
            "\n",
            "\n",
            "To take a quick anonymous survey, run:\n",
            "  $ gcloud alpha survey\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "iGl4SY_tOiJd",
        "colab_type": "code",
        "outputId": "5676508c-6f59-45bf-e506-ad8448eeec04",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 5678
        }
      },
      "source": [
        "!bq help"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Python script for interacting with BigQuery.\n",
            "\n",
            "\n",
            "USAGE: bq [--global_flags] <command> [--command_flags] [args]\n",
            "\n",
            "\n",
            "Any of the following commands:\n",
            "  cancel, cp, extract, head, help, init, insert, load, ls, mk, mkdef, partition,\n",
            "  query, rm, shell, show, update, version, wait\n",
            "\n",
            "\n",
            "cancel     Request a cancel and waits for the job to be cancelled.\n",
            "\n",
            "           Requests a cancel and then either: a) waits until the job is done if\n",
            "           the sync flag is set [default], or b) returns immediately if the sync\n",
            "           flag is not set. Not all job types support a cancel, an error is\n",
            "           returned if it cannot be cancelled. Even for jobs that support a\n",
            "           cancel, success is not guaranteed, the job may have completed by the\n",
            "           time the cancel request is noticed, or the job may be in a stage\n",
            "           where it cannot be cancelled.\n",
            "\n",
            "           Examples:\n",
            "           bq cancel job_id # Requests a cancel and waits until the job is done.\n",
            "           bq --nosync cancel job_id # Requests a cancel and returns\n",
            "           immediately.\n",
            "\n",
            "           Arguments:\n",
            "           job_id: Job ID to cancel.\n",
            "\n",
            "cp         Copies one table to another.\n",
            "\n",
            "           Examples:\n",
            "           bq cp dataset.old_table dataset2.new_table\n",
            "           bq cp --destination_kms_key=kms_key dataset.old_table\n",
            "           dataset2.new_table\n",
            "\n",
            "extract    Perform an extract operation of source_table into destination_uris.\n",
            "\n",
            "           Usage:\n",
            "           extract <source_table> <destination_uris>\n",
            "\n",
            "           Examples:\n",
            "           bq extract ds.summary gs://mybucket/summary.csv\n",
            "\n",
            "           Arguments:\n",
            "           source_table: Source table to extract.\n",
            "           destination_uris: One or more Google Cloud Storage URIs, separated by\n",
            "           commas.\n",
            "\n",
            "head       Displays rows in a table.\n",
            "\n",
            "           Examples:\n",
            "           bq head dataset.table\n",
            "           bq head -j job\n",
            "           bq head -n 10 dataset.table\n",
            "           bq head -s 5 -n 10 dataset.table\n",
            "\n",
            "help       Help for all or selected command:\n",
            "               bq help [<command>]\n",
            "\n",
            "           To retrieve help with global flags:\n",
            "               bq --help\n",
            "\n",
            "           To retrieve help with flags only from the main module:\n",
            "               bq --helpshort [<command>]\n",
            "\n",
            "init       Authenticate and create a default .bigqueryrc file.\n",
            "\n",
            "insert     Inserts rows in a table.\n",
            "\n",
            "           Inserts the records formatted as newline delimited JSON from file\n",
            "           into the specified table. If file is not specified, reads from stdin.\n",
            "           If there were any insert errors it prints the errors to stdout.\n",
            "\n",
            "           Examples:\n",
            "           bq insert dataset.table /tmp/mydata.json\n",
            "           echo '{\"a\":1, \"b\":2}' | bq insert dataset.table\n",
            "\n",
            "           Template table examples: Insert to dataset.template_suffix table\n",
            "           using dataset.template table as its template.\n",
            "           bq insert -x=_suffix dataset.table /tmp/mydata.json\n",
            "\n",
            "load       Perform a load operation of source into destination_table.\n",
            "\n",
            "           Usage:\n",
            "           load <destination_table> <source> [<schema>]\n",
            "\n",
            "           The <destination_table> is the fully-qualified table name of table to\n",
            "           create, or append to if the table already exists.\n",
            "\n",
            "           The <source> argument can be a path to a single local file, or a\n",
            "           comma-separated list of URIs.\n",
            "\n",
            "           The <schema> argument should be either the name of a JSON file or a\n",
            "           text schema. This schema should be omitted if the table already has\n",
            "           one.\n",
            "\n",
            "           In the case that the schema is provided in text form, it should be a\n",
            "           comma-separated list of entries of the form name[:type], where type\n",
            "           will default to string if not specified.\n",
            "\n",
            "           In the case that <schema> is a filename, it should contain a single\n",
            "           array object, each entry of which should be an object with properties\n",
            "           'name', 'type', and (optionally) 'mode'. See the online documentation\n",
            "           for more detail:\n",
            "           https://developers.google.com/bigquery/preparing-data-for-bigquery\n",
            "\n",
            "           Note: the case of a single-entry schema with no type specified is\n",
            "           ambiguous; one can use name:string to force interpretation as a\n",
            "           text schema.\n",
            "\n",
            "           Examples:\n",
            "           bq load ds.new_tbl ./info.csv ./info_schema.json\n",
            "           bq load ds.new_tbl gs://mybucket/info.csv ./info_schema.json\n",
            "           bq load ds.small gs://mybucket/small.csv name:integer,value:string\n",
            "           bq load ds.small gs://mybucket/small.csv field1,field2,field3\n",
            "\n",
            "           Arguments:\n",
            "           destination_table: Destination table name.\n",
            "           source: Name of local file to import, or a comma-separated list of\n",
            "           URI paths to data to import.\n",
            "           schema: Either a text schema or JSON file, as above.\n",
            "\n",
            "ls         List the objects contained in the named collection.\n",
            "\n",
            "           List the objects in the named project or dataset. A trailing : or .\n",
            "           can be used to signify a project or dataset.\n",
            "           * With -j, show the jobs in the named project.\n",
            "           * With -p, show all projects.\n",
            "\n",
            "           Examples:\n",
            "           bq ls\n",
            "           bq ls -j proj\n",
            "           bq ls -p -n 1000\n",
            "           bq ls mydataset\n",
            "           bq ls -a\n",
            "           bq ls -m mydataset\n",
            "           bq ls --routines mydataset (requires whitelisting)\n",
            "           bq ls --filter labels.color:red\n",
            "           bq ls --filter 'labels.color:red labels.size:*'\n",
            "           bq ls --transfer_config --transfer_location='us'\n",
            "           --filter='dataSourceIds:play,adwords'\n",
            "           bq ls --transfer_run --filter='states:SUCCESSED,PENDING'\n",
            "           --run_attempt='LATEST' projects/p/locations/l/transferConfigs/c\n",
            "           bq ls --transfer_log --message_type='messageTypes:INFO,ERROR'\n",
            "           projects/p/locations/l/transferConfigs/c/runs/r\n",
            "           bq ls --reservation_grant --project_id=proj --location='us'\n",
            "           bq ls --reservation_grant --project_id=proj --location='us'\n",
            "           --reservation\n",
            "           <reservation_ref>\n",
            "           bq ls --connection --project_id=proj --location=us\n",
            "\n",
            "mk         Create a dataset, table, view, or transfer configuration with this\n",
            "           name.\n",
            "\n",
            "           See 'bq help load' for more information on specifying the schema.\n",
            "\n",
            "           Examples:\n",
            "           bq mk new_dataset\n",
            "           bq mk new_dataset.new_table\n",
            "           bq --dataset_id=new_dataset mk table\n",
            "           bq mk -t new_dataset.newtable name:integer,value:string\n",
            "           bq mk --view='select 1 as num' new_dataset.newview\n",
            "           (--view_udf_resource=path/to/file.js)\n",
            "           bq mk --materialized_view='select sum(x) as sum_x from dataset.table'\n",
            "           new_dataset.newview\n",
            "           bq mk -d --data_location=EU new_dataset\n",
            "           bq mk --transfer_config --target_dataset=dataset --display_name=name\n",
            "           -p='{\"param\":\"value\"}' --data_source=source\n",
            "           --schedule_start_time={schedule_start_time}\n",
            "           --schedule_end_time={schedule_end_time}\n",
            "           bq mk --transfer_run --start_time={start_time} --end_time={end_time}\n",
            "           projects/p/locations/l/transferConfigs/c\n",
            "           bq mk ---reservation_grant -project_id=proj --location=us\n",
            "           --reservation_id=project:us.dev --job_type=QUERY\n",
            "           bq mk --connection --connection_type='CLOUD_SQL'\n",
            "           --properties='{\"instanceId\" : \"instance\",\n",
            "           \"database\" : \"db\", \"type\" : \"MYSQL\" }'\n",
            "           --connection_credential='{\"username\":\"u\", \"password\":\"p\"}'\n",
            "           --project_id=proj --location=us new_connection\n",
            "\n",
            "mkdef      Emits a definition in JSON for a GCS backed table.\n",
            "\n",
            "           The output of this command can be redirected to a file and used for\n",
            "           the external_table_definition flag with the \"bq query\" and \"bq mk\"\n",
            "           commands. It produces a definition with the most commonly used values\n",
            "           for options. You can modify the output to override option values.\n",
            "\n",
            "           Usage:\n",
            "           mkdef <source_uris> [<schema>]\n",
            "\n",
            "           Examples:\n",
            "           bq mkdef 'gs://bucket/file.csv' field1:integer,field2:string\n",
            "\n",
            "           Arguments:\n",
            "           source_uris: a comma-separated list of uris.\n",
            "           schema: The <schema> argument should be either the name of a JSON\n",
            "           file or\n",
            "           a text schema.\n",
            "\n",
            "           In the case that the schema is provided in text form, it should be a\n",
            "           comma-separated list of entries of the form name[:type], where type\n",
            "           will\n",
            "           default to string if not specified.\n",
            "\n",
            "           In the case that <schema> is a filename, it should contain a\n",
            "           single array object, each entry of which should be an object with\n",
            "           properties 'name', 'type', and (optionally) 'mode'. See the online\n",
            "           documentation for more detail:\n",
            "           https://developers.google.com/bigquery/preparing-data-for-bigquery\n",
            "\n",
            "           Note: the case of a single-entry schema with no type specified is\n",
            "           ambiguous; one can use name:string to force interpretation as a\n",
            "           text schema.\n",
            "\n",
            "partition  Copies source tables into partitioned tables.\n",
            "\n",
            "           Usage: bq partition <source_table_prefix>\n",
            "           <destination_partitioned_table>\n",
            "\n",
            "           Copies tables of the format <source_table_prefix><YYYYmmdd> to a\n",
            "           destination partitioned table, with the date suffix of the source\n",
            "           tables becoming the partition date of the destination table\n",
            "           partitions.\n",
            "\n",
            "           If the destination table does not exist, one will be created with a\n",
            "           schema and that matches the last table that matches the supplied\n",
            "           prefix.\n",
            "\n",
            "           Examples:\n",
            "           bq partition dataset1.sharded_ dataset2.partitioned_table\n",
            "\n",
            "query      Execute a query.\n",
            "\n",
            "           Query should be specifed on command line, or passed on stdin.\n",
            "\n",
            "           Examples:\n",
            "           bq query 'select count(*) from publicdata:samples.shakespeare'\n",
            "           echo 'select count(*) from publicdata:samples.shakespeare' | bq query\n",
            "\n",
            "           Usage:\n",
            "           query [<sql_query>]\n",
            "\n",
            "rm         Delete the dataset, table, or transfer config described by\n",
            "           identifier.\n",
            "\n",
            "           Always requires an identifier, unlike the show and ls commands. By\n",
            "           default, also requires confirmation before deleting. Supports the -d\n",
            "           -t flags to signify that the identifier is a dataset or table.\n",
            "           * With -f, don't ask for confirmation before deleting.\n",
            "           * With -r, remove all tables in the named dataset.\n",
            "\n",
            "           Examples:\n",
            "           bq rm ds.table\n",
            "           bq rm -m ds.model\n",
            "           bq rm --routine ds.routine (requires whitelisting)\n",
            "           bq rm -r -f old_dataset\n",
            "           bq rm --transfer_config=projects/p/locations/l/transferConfigs/c\n",
            "           bq rm --reservation_grant --project_id=proj --location=us\n",
            "           query_proj_dev\n",
            "           bq rm --connection --project_id=proj --location=us con\n",
            "\n",
            "shell      Start an interactive bq session.\n",
            "\n",
            "show       Show all information about an object.\n",
            "\n",
            "           Examples:\n",
            "           bq show -j <job_id>\n",
            "           bq show dataset\n",
            "           bq show [--schema] dataset.table\n",
            "           bq show [--view] dataset.view\n",
            "           bq show [--materialized_view] dataset.materialized_view\n",
            "           bq show -m ds.model\n",
            "           bq show --routine ds.routine (requires whitelisting)\n",
            "           bq show --transfer_config projects/p/locations/l/transferConfigs/c\n",
            "           bq show --transfer_run\n",
            "           projects/p/locations/l/transferConfigs/c/runs/r\n",
            "           bq show --encryption_service_account\n",
            "           bq show --connection --project_id=project --location=us connection\n",
            "\n",
            "update     Updates a dataset, table, view or transfer configuration with this\n",
            "           name.\n",
            "\n",
            "           See 'bq help load' for more information on specifying the schema.\n",
            "\n",
            "           Examples:\n",
            "           bq update --description \"Dataset description\" existing_dataset\n",
            "           bq update --description \"My table\" existing_dataset.existing_table\n",
            "           bq update --description \"My model\" -m existing_dataset.existing_model\n",
            "           bq update -t existing_dataset.existing_table\n",
            "           name:integer,value:string\n",
            "           bq update --destination_kms_key\n",
            "           projects/p/locations/l/keyRings/r/cryptoKeys/k\n",
            "           existing_dataset.existing_table\n",
            "           bq update --view='select 1 as num' existing_dataset.existing_view\n",
            "           (--view_udf_resource=path/to/file.js)\n",
            "           bq update --transfer_config --display_name=name\n",
            "           -p='{\"param\":\"value\"}'\n",
            "           projects/p/locations/l/transferConfigs/c\n",
            "           bq update --transfer_config --target_dataset=dataset\n",
            "           --refresh_window_days=5 --update_credentials\n",
            "           projects/p/locations/l/transferConfigs/c\n",
            "           bq update --reservation --location=US --project_id=my-project\n",
            "           --reservation_size=2G\n",
            "           bq update --connection_credential='{\"username\":\"u\", \"password\":\"p\"}'\n",
            "           --location=US --project_id=my-project existing_connection\n",
            "\n",
            "version    Return the version of bq.\n",
            "\n",
            "wait       Wait some number of seconds for a job to finish.\n",
            "\n",
            "           Poll job_id until either (1) the job is DONE or (2) the specified\n",
            "           number of seconds have elapsed. Waits forever if unspecified. If no\n",
            "           job_id is specified, and there is only one running job, we poll that\n",
            "           job.\n",
            "\n",
            "           Examples:\n",
            "           bq wait # Waits forever for the currently running job.\n",
            "           bq wait job_id # Waits forever\n",
            "           bq wait job_id 100 # Waits 100 seconds\n",
            "           bq wait job_id 0 # Polls if a job is done, then returns immediately.\n",
            "           # These may exit with a non-zero status code to indicate \"failure\":\n",
            "           bq wait --fail_on_error job_id # Succeeds if job succeeds.\n",
            "           bq wait --fail_on_error job_id 100 # Succeeds if job succeeds in 100\n",
            "           sec.\n",
            "\n",
            "           Arguments:\n",
            "           job_id: Job ID to wait on.\n",
            "           secs: Number of seconds to wait (must be >= 0).\n",
            "\n",
            "\n",
            "Run 'bq --help' to get help for global flags.\n",
            "Run 'bq help <command>' to get help for <command>.\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "mJHpzlkwOyXO",
        "colab_type": "code",
        "outputId": "b326494e-80db-4d2e-a8bb-5307237fed5b",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 445
        }
      },
      "source": [
        "!gcloud auth application-default login"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "\n",
            "The environment variable [GOOGLE_APPLICATION_CREDENTIALS] is set to:\n",
            "  [/content/adc.json]\n",
            "Credentials will still be generated to the default location:\n",
            "  [/content/.config/application_default_credentials.json]\n",
            "To use these credentials, unset this environment variable before\n",
            "running your application.\n",
            "\n",
            "Do you want to continue (Y/n)?  Y\n",
            "\n",
            "Go to the following link in your browser:\n",
            "\n",
            "    https://accounts.google.com/o/oauth2/auth?redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&prompt=select_account&response_type=code&client_id=764086051850-6qr4p6gpi6hn506pt8ejuq83di341hur.apps.googleusercontent.com&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.email+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fcloud-platform&access_type=offline\n",
            "\n",
            "\n",
            "Enter verification code: 4/WgHQ4Gey91fMt_RCzRObQFOASe58WAJOSPP0qJL-aFms-c9vIZ4PmZo\n",
            "\n",
            "Credentials saved to file: [/content/.config/application_default_credentials.json]\n",
            "\n",
            "These credentials will be used by any library that requests\n",
            "Application Default Credentials.\n",
            "\n",
            "To generate an access token for other uses, run:\n",
            "  gcloud auth application-default print-access-token\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "__SsUc8K9laj",
        "colab_type": "code",
        "outputId": "c814fdfa-71bc-4a0e-d697-38dfaf5f5d82",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 6973
        }
      },
      "source": [
        "!bq --project_id isb-cgc show \"bigquery-public-data:human_variant_annotation.gnomad_gnomad_exomes_hg19_release_2_0_2\""
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "\n",
            "Welcome to BigQuery! This script will walk you through the \n",
            "process of initializing your .bigqueryrc configuration file.\n",
            "\n",
            "First, we need to set up your credentials if they do not \n",
            "already exist.\n",
            "\n",
            "Credential creation complete. Now we will select a default project.\n",
            "\n",
            "List of projects:\n",
            "   #             projectId                     friendlyName           \n",
            " ----- ----------------------------- -------------------------------- \n",
            "  1     another-test-project-186620   another-test-project            \n",
            "  2     ultra-envoy-227421            APT Sandbox                     \n",
            "  3     isb-cgc-01-0010               B-cell Receptor Analysis        \n",
            "  4     big-data-science              big-data-science                \n",
            "  5     cgc-05-0042                   biocellion                      \n",
            "  6     cancer-data-commons           cancer-data-commons             \n",
            "  7     cancer-genome-atlas           cancer-genome-atlas             \n",
            "  8     cgc-05-0001                   CGC-05-0001                     \n",
            "  9     cgc-05-0002                   CGC-05-0002                     \n",
            "  10    cgc-05-0004                   CGC-05-0004                     \n",
            "  11    cgc-05-0006                   CGC-05-0006                     \n",
            "  12    cgc-05-0007                   CGC-05-0007                     \n",
            "  13    cgc-05-0009                   CGC-05-0009                     \n",
            "  14    cgc-05-0010                   CGC-05-0010                     \n",
            "  15    cgc-05-0015                   CGC-05-0015                     \n",
            "  16    cgc-05-0021                   CGC-05-0021                     \n",
            "  17    cgc-05-0022                   CGC-05-0022                     \n",
            "  18    cgc-05-0026                   CGC-05-0026                     \n",
            "  19    cgc-05-0027                   CGC-05-0027                     \n",
            "  20    cgc-05-0031                   CGC-05-0031                     \n",
            "  21    cgc-05-0032                   CGC-05-0032                     \n",
            "  22    cgc-05-0033                   CGC-05-0033                     \n",
            "  23    cgc-05-0036                   CGC-05-0036                     \n",
            "  24    cgc-05-0037                   CGC-05-0037                     \n",
            "  25    cgc-05-0038                   CGC-05-0038                     \n",
            "  26    cgc-05-0040                   CGC-05-0040--0                  \n",
            "  27    cgc-05-0041                   CGC-05-0041                     \n",
            "  28    cgc-05-0050                   CGC-05-0050 - internal-invoice  \n",
            "  29    cgc-05-0051                   CGC-05-0051-external-invoice    \n",
            "  30    chc-nih-chest-xray            chc-nih-chest-xray              \n",
            "  31    chc-tcia                      chc-tcia                        \n",
            "  32    isb-cgc-04-0003               COSMIC                          \n",
            "  33    csra-test-project             csra-test-project               \n",
            "  34    csra-testing02                csra-testing02                  \n",
            "  35    dceg-connect                  dceg-connect                    \n",
            "  36    dcf-prod-buckets              dcf-prod-buckets                \n",
            "  37    dcf-prod-manifests            dcf-prod-manifests              \n",
            "  38    dcf-staging-202214            dcf-staging                     \n",
            "  39    digital-pathology-imaging     digital-pathology-imaging       \n",
            "  40    digital-radiology-imaging     digital-radiology-imaging       \n",
            "  41    fc-product-demo               fc-product-demo                 \n",
            "  42    g-doc-pilot                   g-doc-pilot                     \n",
            "  43    cgc-05-0017                   GDC ISB collaboration           \n",
            "  44    isb-cgc-04-0002               GDSC                            \n",
            "  45    gm-kids-first                 gm-kids-first                   \n",
            "  46    gmkf-ews                      gmkf-ews                        \n",
            "  47    htan-dcc                      htan-dcc                        \n",
            "  48    htan-dcc-isb-01               htan-dcc-isb-01                 \n",
            "  49    cgc-05-0035                   iAtlas project                  \n",
            "  50    imaging-data-commons          imaging-data-commons            \n",
            "  51    isb-cgc-01-0007               Immune Response - PanCanAtlas   \n",
            "  52    isb-cgc-04-0008               ISB GBM project                 \n",
            "  53    isb-cgc                       ISB-CGC                         \n",
            "  54    isb-cgc-04-0030               ISB-CGC workshop                \n",
            "  55    isb-cgc-01-0002               ISB-CGC-01-0002                 \n",
            "  56    isb-cgc-01-0005               ISB-CGC-01-0005                 \n",
            "  57    isb-cgc-01-0006               ISB-CGC-01-0006                 \n",
            "  58    isb-cgc-01-0009               ISB-CGC-01-0009                 \n",
            "  59    isb-cgc-02-0001               ISB-CGC-02-0001                 \n",
            "  60    isb-cgc-03-0001               ISB-CGC-03-0001                 \n",
            "  61    isb-cgc-03-0002               ISB-CGC-03-0002                 \n",
            "  62    isb-cgc-03-0004               ISB-CGC-03-0004                 \n",
            "  63    isb-cgc-03-0006               ISB-CGC-03-0006                 \n",
            "  64    isb-cgc-04-0007               ISB-CGC-04-0007                 \n",
            "  65    isb-cgc-04-0013               ISB-CGC-04-0013                 \n",
            "  66    isb-cgc-04-0017               ISB-CGC-04-0017                 \n",
            "  67    isb-cgc-04-0018               ISB-CGC-04-0018                 \n",
            "  68    isb-cgc-04-0021               ISB-CGC-04-0021                 \n",
            "  69    isb-cgc-04-0022               ISB-CGC-04-0022                 \n",
            "  70    isb-cgc-04-0023               ISB-CGC-04-0023                 \n",
            "  71    isb-cgc-04-0024               ISB-CGC-04-0024                 \n",
            "  72    isb-cgc-04-0026               ISB-CGC-04-0026                 \n",
            "  73    isb-cgc-04-0027               ISB-CGC-04-0027                 \n",
            "  74    isb-cgc-06-0005               ISB-CGC-06-0005                 \n",
            "  75    isb-cgc-06-0010               ISB-CGC-06-0010                 \n",
            "  76    isb-cgc-06-0011               ISB-CGC-06-0011                 \n",
            "  77    isb-cgc-admin                 ISB-CGC-Admin                   \n",
            "  78    isb-cgc-bq                    isb-cgc-bq                      \n",
            "  79    isb-cgc-cbq                   isb-cgc-cbq                     \n",
            "  80    isb-cgc-cosmic                isb-cgc-cosmic                  \n",
            "  81    isb-cgc-data-01               ISB-CGC-Data-01                 \n",
            "  82    isb-cgc-data-02               ISB-CGC-Data-02                 \n",
            "  83    isb-cgc-data-open             isb-cgc-data-open               \n",
            "  84    isb-cgc-ivy-gap               isb-cgc-ivy-gap                 \n",
            "  85    isb-cgc-kf                    isb-cgc-kf                      \n",
            "  86    isb-cgc-obq                   isb-cgc-obq                     \n",
            "  87    isb-cgc-production            ISB-CGC-production              \n",
            "  88    isb-cgc-secure-testing        isb-cgc-secure-testing          \n",
            "  89    isb-cgc-stage                 ISB-CGC-stage                   \n",
            "  90    isb-cgc-test                  ISB-CGC-test                    \n",
            "  91    isb-cgc-uat                   ISB-CGC-UAT                     \n",
            "  92    isb-project-zero              isb-project-zero                \n",
            "  93    isb-tcga-phs000178-test       isb-tcga-phs000178-test         \n",
            "  94    isl-resource                  iSL-resource                    \n",
            "  95    cgc-05-0034                   ISPY                            \n",
            "  96    kids-first-drc                kids-first-drc                  \n",
            "  97    cgc-05-0014                   MC3-OxoG                        \n",
            "  98    isb-cgc-04-0010               miRNA project                   \n",
            "  99    neural-journey-186022         My First Project                \n",
            "  100   ncats-translator              NCATS-Translator                \n",
            "  101   nci-cancer-data-commons       nci-cancer-data-commons         \n",
            "  102   nci-image-data-commons        nci-image-data-commons          \n",
            "  103   nci-imaging-data-commons      nci-imaging-data-commons        \n",
            "  104   nci-research-data-commons     nci-research-data-commons       \n",
            "  105   new-june2018-project          new-june2018-project            \n",
            "  106   nih-data-commons              nih-data-commons                \n",
            "  107   isb-cgc-01-0008               PanCanAtlas Data Mirror         \n",
            "  108   pancancer-atlas               pancancer-atlas                 \n",
            "  109   cgc-05-0012                   PanGastrointestinal             \n",
            "  110   cgc-05-0003                   Pathology Imaging               \n",
            "  111   isb-cgc-04-0025               PCA Fusion Analysis             \n",
            "  112   isb-cgc-06-0004               PCA germline AWG                \n",
            "  113   isb-cgc-04-0028               PCA RNAseq                      \n",
            "  114   isb-cgc-01-0003               PCA sandbox                     \n",
            "  115   cgc-05-0030                   QIN HEADNECK                    \n",
            "  116   cgc-05-0011                   Radiology Imaging               \n",
            "  117   research-data-commons         research-data-commons           \n",
            "  118   cgc-05-0029                   Sage on ISB-CGC                 \n",
            "  119   isb-cgc-04-0029               smc-rna-eval                    \n",
            "  120   tcga-qc                       TCGA-QC                         \n",
            "  121   test-project-in-isb-cgc       test project in isb cgc         \n",
            "  122   uw-isb-p01                    uw-isb-p01                      \n",
            "  123   wjrl-project-thing            wjrl-project-thing              \n",
            "Found multiple projects. Please enter a selection for \n",
            "which should be the default, or leave blank to not \n",
            "set a default.\n",
            "\n",
            "Enter a selection (1 - 123): 78\n",
            "\n",
            "BigQuery configuration complete! Type \"bq\" to get started.\n",
            "\n",
            "Table bigquery-public-data:human_variant_annotation.gnomad_gnomad_exomes_hg19_release_2_0_2\n",
            "\n",
            "   Last modified                      Schema                      Total Rows   Total Bytes   Expiration               Time Partitioning               Labels  \n",
            " ----------------- --------------------------------------------- ------------ ------------- ------------ ------------------------------------------- -------- \n",
            "  05 Sep 14:45:28   |- reference_name: string                     15014473     72556649388                DAY (field: partition_date_please_ignore)           \n",
            "                    |- start_position: integer                                                                                                                \n",
            "                    |- end_position: integer                                                                                                                  \n",
            "                    |- reference_bases: string                                                                                                                \n",
            "                    +- alternate_bases: record (repeated)                                                                                                     \n",
            "                    |  |- alt: string                                                                                                                         \n",
            "                    |  |- AC: integer                                                                                                                         \n",
            "                    |  |- AF: float                                                                                                                           \n",
            "                    |  |- GQ_HIST_ALT: string                                                                                                                 \n",
            "                    |  |- DP_HIST_ALT: string                                                                                                                 \n",
            "                    |  |- AB_HIST_ALT: string                                                                                                                 \n",
            "                    |  |- AC_AFR: integer                                                                                                                     \n",
            "                    |  |- AC_AMR: integer                                                                                                                     \n",
            "                    |  |- AC_ASJ: integer                                                                                                                     \n",
            "                    |  |- AC_EAS: integer                                                                                                                     \n",
            "                    |  |- AC_FIN: integer                                                                                                                     \n",
            "                    |  |- AC_NFE: integer                                                                                                                     \n",
            "                    |  |- AC_OTH: integer                                                                                                                     \n",
            "                    |  |- AC_SAS: integer                                                                                                                     \n",
            "                    |  |- AC_Male: integer                                                                                                                    \n",
            "                    |  |- AC_Female: integer                                                                                                                  \n",
            "                    |  |- AF_AFR: float                                                                                                                       \n",
            "                    |  |- AF_AMR: float                                                                                                                       \n",
            "                    |  |- AF_ASJ: float                                                                                                                       \n",
            "                    |  |- AF_EAS: float                                                                                                                       \n",
            "                    |  |- AF_FIN: float                                                                                                                       \n",
            "                    |  |- AF_NFE: float                                                                                                                       \n",
            "                    |  |- AF_OTH: float                                                                                                                       \n",
            "                    |  |- AF_SAS: float                                                                                                                       \n",
            "                    |  |- AF_Male: float                                                                                                                      \n",
            "                    |  |- AF_Female: float                                                                                                                    \n",
            "                    |  |- AC_raw: integer                                                                                                                     \n",
            "                    |  |- AF_raw: float                                                                                                                       \n",
            "                    |  |- Hom_AFR: integer                                                                                                                    \n",
            "                    |  |- Hom_AMR: integer                                                                                                                    \n",
            "                    |  |- Hom_ASJ: integer                                                                                                                    \n",
            "                    |  |- Hom_EAS: integer                                                                                                                    \n",
            "                    |  |- Hom_FIN: integer                                                                                                                    \n",
            "                    |  |- Hom_NFE: integer                                                                                                                    \n",
            "                    |  |- Hom_OTH: integer                                                                                                                    \n",
            "                    |  |- Hom_SAS: integer                                                                                                                    \n",
            "                    |  |- Hom_Male: integer                                                                                                                   \n",
            "                    |  |- Hom_Female: integer                                                                                                                 \n",
            "                    |  |- Hom_raw: integer                                                                                                                    \n",
            "                    |  |- Hom: integer                                                                                                                        \n",
            "                    |  |- POPMAX: string                                                                                                                      \n",
            "                    |  |- AC_POPMAX: integer                                                                                                                  \n",
            "                    |  |- AN_POPMAX: integer                                                                                                                  \n",
            "                    |  |- AF_POPMAX: float                                                                                                                    \n",
            "                    |  |- DP_MEDIAN: integer                                                                                                                  \n",
            "                    |  |- DREF_MEDIAN: float                                                                                                                  \n",
            "                    |  |- GQ_MEDIAN: integer                                                                                                                  \n",
            "                    |  |- AB_MEDIAN: float                                                                                                                    \n",
            "                    |  |- AS_RF: float                                                                                                                        \n",
            "                    |  |- AS_FilterStatus: string                                                                                                             \n",
            "                    |  |- AC_AFR_Male: integer                                                                                                                \n",
            "                    |  |- AF_AMR_Male: float                                                                                                                  \n",
            "                    |  |- Hemi_NFE: integer                                                                                                                   \n",
            "                    |  |- Hemi_AFR: integer                                                                                                                   \n",
            "                    |  |- AC_ASJ_Female: integer                                                                                                              \n",
            "                    |  |- AF_FIN_Female: float                                                                                                                \n",
            "                    |  |- AC_OTH_Female: integer                                                                                                              \n",
            "                    |  |- AC_NFE_Female: integer                                                                                                              \n",
            "                    |  |- AC_EAS_Male: integer                                                                                                                \n",
            "                    |  |- AC_OTH_Male: integer                                                                                                                \n",
            "                    |  |- Hemi_AMR: integer                                                                                                                   \n",
            "                    |  |- AC_NFE_Male: integer                                                                                                                \n",
            "                    |  |- Hemi: integer                                                                                                                       \n",
            "                    |  |- Hemi_SAS: integer                                                                                                                   \n",
            "                    |  |- AF_FIN_Male: float                                                                                                                  \n",
            "                    |  |- AF_AFR_Male: float                                                                                                                  \n",
            "                    |  |- AF_SAS_Male: float                                                                                                                  \n",
            "                    |  |- Hemi_ASJ: integer                                                                                                                   \n",
            "                    |  |- Hemi_raw: integer                                                                                                                   \n",
            "                    |  |- AF_OTH_Male: float                                                                                                                  \n",
            "                    |  |- AC_SAS_Female: integer                                                                                                              \n",
            "                    |  |- AF_NFE_Female: float                                                                                                                \n",
            "                    |  |- AF_EAS_Female: float                                                                                                                \n",
            "                    |  |- AF_EAS_Male: float                                                                                                                  \n",
            "                    |  |- AF_SAS_Female: float                                                                                                                \n",
            "                    |  |- AF_AFR_Female: float                                                                                                                \n",
            "                    |  |- AC_FIN_Female: integer                                                                                                              \n",
            "                    |  |- Hemi_OTH: integer                                                                                                                   \n",
            "                    |  |- AC_AFR_Female: integer                                                                                                              \n",
            "                    |  |- AF_AMR_Female: float                                                                                                                \n",
            "                    |  |- AC_AMR_Male: integer                                                                                                                \n",
            "                    |  |- AC_SAS_Male: integer                                                                                                                \n",
            "                    |  |- AF_ASJ_Male: float                                                                                                                  \n",
            "                    |  |- AC_EAS_Female: integer                                                                                                              \n",
            "                    |  |- AC_AMR_Female: integer                                                                                                              \n",
            "                    |  |- Hemi_FIN: integer                                                                                                                   \n",
            "                    |  |- AC_FIN_Male: integer                                                                                                                \n",
            "                    |  |- AF_ASJ_Female: float                                                                                                                \n",
            "                    |  |- AF_OTH_Female: float                                                                                                                \n",
            "                    |  |- AF_NFE_Male: float                                                                                                                  \n",
            "                    |  |- AC_ASJ_Male: integer                                                                                                                \n",
            "                    |  |- Hemi_EAS: integer                                                                                                                   \n",
            "                    |  +- CSQ: record (repeated)                                                                                                              \n",
            "                    |  |  |- allele: string                                                                                                                   \n",
            "                    |  |  |- Consequence: string                                                                                                              \n",
            "                    |  |  |- IMPACT: string                                                                                                                   \n",
            "                    |  |  |- SYMBOL: string                                                                                                                   \n",
            "                    |  |  |- Gene: string                                                                                                                     \n",
            "                    |  |  |- Feature_type: string                                                                                                             \n",
            "                    |  |  |- Feature: string                                                                                                                  \n",
            "                    |  |  |- BIOTYPE: string                                                                                                                  \n",
            "                    |  |  |- EXON: string                                                                                                                     \n",
            "                    |  |  |- INTRON: string                                                                                                                   \n",
            "                    |  |  |- HGVSc: string                                                                                                                    \n",
            "                    |  |  |- HGVSp: string                                                                                                                    \n",
            "                    |  |  |- cDNA_position: string                                                                                                            \n",
            "                    |  |  |- CDS_position: string                                                                                                             \n",
            "                    |  |  |- Protein_position: string                                                                                                         \n",
            "                    |  |  |- Amino_acids: string                                                                                                              \n",
            "                    |  |  |- Codons: string                                                                                                                   \n",
            "                    |  |  |- Existing_variation: string                                                                                                       \n",
            "                    |  |  |- ALLELE_NUM: string                                                                                                               \n",
            "                    |  |  |- DISTANCE: string                                                                                                                 \n",
            "                    |  |  |- STRAND: string                                                                                                                   \n",
            "                    |  |  |- FLAGS: string                                                                                                                    \n",
            "                    |  |  |- VARIANT_CLASS: string                                                                                                            \n",
            "                    |  |  |- MINIMISED: string                                                                                                                \n",
            "                    |  |  |- SYMBOL_SOURCE: string                                                                                                            \n",
            "                    |  |  |- HGNC_ID: string                                                                                                                  \n",
            "                    |  |  |- CANONICAL: string                                                                                                                \n",
            "                    |  |  |- TSL: string                                                                                                                      \n",
            "                    |  |  |- APPRIS: string                                                                                                                   \n",
            "                    |  |  |- CCDS: string                                                                                                                     \n",
            "                    |  |  |- ENSP: string                                                                                                                     \n",
            "                    |  |  |- SWISSPROT: string                                                                                                                \n",
            "                    |  |  |- TREMBL: string                                                                                                                   \n",
            "                    |  |  |- UNIPARC: string                                                                                                                  \n",
            "                    |  |  |- GENE_PHENO: string                                                                                                               \n",
            "                    |  |  |- SIFT: string                                                                                                                     \n",
            "                    |  |  |- PolyPhen: string                                                                                                                 \n",
            "                    |  |  |- DOMAINS: string                                                                                                                  \n",
            "                    |  |  |- HGVS_OFFSET: string                                                                                                              \n",
            "                    |  |  |- GMAF: string                                                                                                                     \n",
            "                    |  |  |- AFR_MAF: string                                                                                                                  \n",
            "                    |  |  |- AMR_MAF: string                                                                                                                  \n",
            "                    |  |  |- EAS_MAF: string                                                                                                                  \n",
            "                    |  |  |- EUR_MAF: string                                                                                                                  \n",
            "                    |  |  |- SAS_MAF: string                                                                                                                  \n",
            "                    |  |  |- AA_MAF: string                                                                                                                   \n",
            "                    |  |  |- EA_MAF: string                                                                                                                   \n",
            "                    |  |  |- ExAC_MAF: string                                                                                                                 \n",
            "                    |  |  |- ExAC_Adj_MAF: string                                                                                                             \n",
            "                    |  |  |- ExAC_AFR_MAF: string                                                                                                             \n",
            "                    |  |  |- ExAC_AMR_MAF: string                                                                                                             \n",
            "                    |  |  |- ExAC_EAS_MAF: string                                                                                                             \n",
            "                    |  |  |- ExAC_FIN_MAF: string                                                                                                             \n",
            "                    |  |  |- ExAC_NFE_MAF: string                                                                                                             \n",
            "                    |  |  |- ExAC_OTH_MAF: string                                                                                                             \n",
            "                    |  |  |- ExAC_SAS_MAF: string                                                                                                             \n",
            "                    |  |  |- CLIN_SIG: string                                                                                                                 \n",
            "                    |  |  |- SOMATIC: string                                                                                                                  \n",
            "                    |  |  |- PHENO: string                                                                                                                    \n",
            "                    |  |  |- PUBMED: string                                                                                                                   \n",
            "                    |  |  |- MOTIF_NAME: string                                                                                                               \n",
            "                    |  |  |- MOTIF_POS: string                                                                                                                \n",
            "                    |  |  |- HIGH_INF_POS: string                                                                                                             \n",
            "                    |  |  |- MOTIF_SCORE_CHANGE: string                                                                                                       \n",
            "                    |  |  |- LoF: string                                                                                                                      \n",
            "                    |  |  |- LoF_filter: string                                                                                                               \n",
            "                    |  |  |- LoF_flags: string                                                                                                                \n",
            "                    |  |  |- LoF_info: string                                                                                                                 \n",
            "                    |- names: string (repeated)                                                                                                               \n",
            "                    |- quality: float                                                                                                                         \n",
            "                    |- filter: string (repeated)                                                                                                              \n",
            "                    +- call: record (repeated)                                                                                                                \n",
            "                    |  |- name: string                                                                                                                        \n",
            "                    |  |- genotype: integer (repeated)                                                                                                        \n",
            "                    |  |- phaseset: string                                                                                                                    \n",
            "                    |  |- AD: integer (repeated)                                                                                                              \n",
            "                    |  |- DP: integer                                                                                                                         \n",
            "                    |  |- GQ: integer                                                                                                                         \n",
            "                    |  |- PL: integer (repeated)                                                                                                              \n",
            "                    |- AN: integer                                                                                                                            \n",
            "                    |- BaseQRankSum: float                                                                                                                    \n",
            "                    |- ClippingRankSum: float                                                                                                                 \n",
            "                    |- DB: boolean                                                                                                                            \n",
            "                    |- DP: integer                                                                                                                            \n",
            "                    |- FS: float                                                                                                                              \n",
            "                    |- InbreedingCoeff: float                                                                                                                 \n",
            "                    |- MQ: float                                                                                                                              \n",
            "                    |- MQRankSum: float                                                                                                                       \n",
            "                    |- QD: float                                                                                                                              \n",
            "                    |- ReadPosRankSum: float                                                                                                                  \n",
            "                    |- SOR: float                                                                                                                             \n",
            "                    |- VQSLOD: float                                                                                                                          \n",
            "                    |- VQSR_culprit: string                                                                                                                   \n",
            "                    |- VQSR_NEGATIVE_TRAIN_SITE: boolean                                                                                                      \n",
            "                    |- VQSR_POSITIVE_TRAIN_SITE: boolean                                                                                                      \n",
            "                    |- GQ_HIST_ALL: string                                                                                                                    \n",
            "                    |- DP_HIST_ALL: string                                                                                                                    \n",
            "                    |- AB_HIST_ALL: string                                                                                                                    \n",
            "                    |- AN_AFR: integer                                                                                                                        \n",
            "                    |- AN_AMR: integer                                                                                                                        \n",
            "                    |- AN_ASJ: integer                                                                                                                        \n",
            "                    |- AN_EAS: integer                                                                                                                        \n",
            "                    |- AN_FIN: integer                                                                                                                        \n",
            "                    |- AN_NFE: integer                                                                                                                        \n",
            "                    |- AN_OTH: integer                                                                                                                        \n",
            "                    |- AN_SAS: integer                                                                                                                        \n",
            "                    |- AN_Male: integer                                                                                                                       \n",
            "                    |- AN_Female: integer                                                                                                                     \n",
            "                    |- GC_AFR: integer (repeated)                                                                                                             \n",
            "                    |- GC_AMR: integer (repeated)                                                                                                             \n",
            "                    |- GC_ASJ: integer (repeated)                                                                                                             \n",
            "                    |- GC_EAS: integer (repeated)                                                                                                             \n",
            "                    |- GC_FIN: integer (repeated)                                                                                                             \n",
            "                    |- GC_NFE: integer (repeated)                                                                                                             \n",
            "                    |- GC_OTH: integer (repeated)                                                                                                             \n",
            "                    |- GC_SAS: integer (repeated)                                                                                                             \n",
            "                    |- GC_Male: integer (repeated)                                                                                                            \n",
            "                    |- GC_Female: integer (repeated)                                                                                                          \n",
            "                    |- AN_raw: integer                                                                                                                        \n",
            "                    |- GC_raw: integer (repeated)                                                                                                             \n",
            "                    |- GC: integer (repeated)                                                                                                                 \n",
            "                    |- STAR_AC: integer                                                                                                                       \n",
            "                    |- STAR_AC_raw: integer                                                                                                                   \n",
            "                    |- STAR_Hom: integer                                                                                                                      \n",
            "                    |- AS_RF_POSITIVE_TRAIN: integer (repeated)                                                                                               \n",
            "                    |- AS_RF_NEGATIVE_TRAIN: integer (repeated)                                                                                               \n",
            "                    |- AN_FIN_Male: integer                                                                                                                   \n",
            "                    |- AN_EAS_Female: integer                                                                                                                 \n",
            "                    |- AN_NFE_Female: integer                                                                                                                 \n",
            "                    |- AN_AMR_Female: integer                                                                                                                 \n",
            "                    |- AN_ASJ_Male: integer                                                                                                                   \n",
            "                    |- GC_OTH_Male: integer (repeated)                                                                                                        \n",
            "                    |- GC_FIN_Male: integer (repeated)                                                                                                        \n",
            "                    |- GC_SAS_Male: integer (repeated)                                                                                                        \n",
            "                    |- AN_FIN_Female: integer                                                                                                                 \n",
            "                    |- GC_EAS_Male: integer (repeated)                                                                                                        \n",
            "                    |- GC_ASJ_Female: integer (repeated)                                                                                                      \n",
            "                    |- GC_SAS_Female: integer (repeated)                                                                                                      \n",
            "                    |- GC_ASJ_Male: integer (repeated)                                                                                                        \n",
            "                    |- AN_ASJ_Female: integer                                                                                                                 \n",
            "                    |- AN_OTH_Male: integer                                                                                                                   \n",
            "                    |- STAR_Hemi: integer                                                                                                                     \n",
            "                    |- AN_SAS_Female: integer                                                                                                                 \n",
            "                    |- AN_AFR_Female: integer                                                                                                                 \n",
            "                    |- AN_OTH_Female: integer                                                                                                                 \n",
            "                    |- GC_AFR_Female: integer (repeated)                                                                                                      \n",
            "                    |- GC_AMR_Male: integer (repeated)                                                                                                        \n",
            "                    |- GC_NFE_Male: integer (repeated)                                                                                                        \n",
            "                    |- GC_NFE_Female: integer (repeated)                                                                                                      \n",
            "                    |- AN_AFR_Male: integer                                                                                                                   \n",
            "                    |- AN_NFE_Male: integer                                                                                                                   \n",
            "                    |- GC_AMR_Female: integer (repeated)                                                                                                      \n",
            "                    |- GC_FIN_Female: integer (repeated)                                                                                                      \n",
            "                    |- GC_EAS_Female: integer (repeated)                                                                                                      \n",
            "                    |- GC_AFR_Male: integer (repeated)                                                                                                        \n",
            "                    |- AN_SAS_Male: integer                                                                                                                   \n",
            "                    |- AN_EAS_Male: integer                                                                                                                   \n",
            "                    |- AN_AMR_Male: integer                                                                                                                   \n",
            "                    |- GC_OTH_Female: integer (repeated)                                                                                                      \n",
            "                    |- segdup: boolean                                                                                                                        \n",
            "                    |- lcr: boolean                                                                                                                           \n",
            "                    |- partition_date_please_ignore: date                                                                                                     \n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "l5BuBSRvCkZy",
        "colab_type": "text"
      },
      "source": [
        "As we can see above, this table has well  over 200 fields!\n",
        "- 124 integer fields (33 of which are *repeated* fields)\n",
        "-  84 string fields (2 of which are *repeated* fields)\n",
        "-  43 floating point fields\n",
        "-  5 boolean fields\n",
        "\n",
        "Some of these fields are contained within 3 (potentially repeated) records: **alternate_bases**, **alternate_bases.CSQ**, and **call**.\n",
        "\n",
        "Note that fully running the **exploreFieldContents** on this table would take several hours and would also run out of RAM prior to completing, so instead we will run it with most fields *excluded*."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "eQV8HS5xqr86",
        "colab_type": "code",
        "outputId": "eadececc-593e-40bc-b86e-f42cb373b078",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1873
        }
      },
      "source": [
        "excludedNames = [ \"start_position\", \"end_position\" ]\n",
        "excludedTypes = [ \"INTEGER\", \"FLOAT\" ]\n",
        "xf = bbq.bbqExploreFieldContents ( bqclient,\n",
        "                                projectName, datasetName, tableName, \n",
        "                                excludedNames, excludedTypes )"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "reference_name                STRING      NULLABLE   (24, 0, 15014473, '1', 1494265, 0.09952164155212108, '1', 'Y', 8, 18)\n",
            "start_position                INTEGER     NULLABLE   (this field was excluded)\n",
            "end_position                  INTEGER     NULLABLE   (this field was excluded)\n",
            "reference_bases               STRING      NULLABLE   (106583, 0, 15014473, 'G', 4617665, 0.30754759091444633, 'A', 'TTTTTTTTTTTTTTTTTTTTTTTTG', 2, 4)\n",
            "alternate_bases               RECORD      REPEATED   [93] \n",
            "    > alt                     STRING      NULLABLE   (133493, 0, 17009588, 'A', 4857970, 0.28560186172645685, 'A', 'TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTG', 2, 4)\n",
            "    > AC                      INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AF                      FLOAT       NULLABLE   (this field was excluded)\n",
            "    > GQ_HIST_ALT             STRING      NULLABLE   (1534167, 0, 17009588, '0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|1', 6471723, 0.3804750003351051, '0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|1', '9|9|9|9|9|22|19|15|15|32|14|13|28|30|17|29|22|22|17|718', 3, 35363)\n",
            "    > DP_HIST_ALT             STRING      NULLABLE   (2562090, 0, 17009588, '0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0', 1994850, 0.11727797287035995, '0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0', '9|9|9|9|2|2|0|0|1|0|0|1|0|0|0|0|0|0|0|0', 23, 861131)\n",
            "    > AB_HIST_ALT             STRING      NULLABLE   (2380437, 6804, 17002784, '0|0|0|0|0|0|0|0|0|1|0|0|0|0|0|0|0|0|0|0', 1542460, 0.0907180847560023, '0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0|0', '9|9|7|4|1|2|1|0|0|0|0|0|0|0|0|0|0|0|0|0', 26, 680158)\n",
            "    > AC_AFR                  INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AC_AMR                  INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AC_ASJ                  INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AC_EAS                  INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AC_FIN                  INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AC_NFE                  INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AC_OTH                  INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AC_SAS                  INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AC_Male                 INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AC_Female               INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AF_AFR                  FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AF_AMR                  FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AF_ASJ                  FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AF_EAS                  FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AF_FIN                  FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AF_NFE                  FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AF_OTH                  FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AF_SAS                  FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AF_Male                 FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AF_Female               FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AC_raw                  INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AF_raw                  FLOAT       NULLABLE   (this field was excluded)\n",
            "    > Hom_AFR                 INTEGER     NULLABLE   (this field was excluded)\n",
            "    > Hom_AMR                 INTEGER     NULLABLE   (this field was excluded)\n",
            "    > Hom_ASJ                 INTEGER     NULLABLE   (this field was excluded)\n",
            "    > Hom_EAS                 INTEGER     NULLABLE   (this field was excluded)\n",
            "    > Hom_FIN                 INTEGER     NULLABLE   (this field was excluded)\n",
            "    > Hom_NFE                 INTEGER     NULLABLE   (this field was excluded)\n",
            "    > Hom_OTH                 INTEGER     NULLABLE   (this field was excluded)\n",
            "    > Hom_SAS                 INTEGER     NULLABLE   (this field was excluded)\n",
            "    > Hom_Male                INTEGER     NULLABLE   (this field was excluded)\n",
            "    > Hom_Female              INTEGER     NULLABLE   (this field was excluded)\n",
            "    > Hom_raw                 INTEGER     NULLABLE   (this field was excluded)\n",
            "    > Hom                     INTEGER     NULLABLE   (this field was excluded)\n",
            "    > POPMAX                  STRING      NULLABLE   (7, 1755119, 15254469, 'NFE', 5324135, 0.34902132614383363, 'AFR', 'SAS', 2, 5)\n",
            "    > AC_POPMAX               INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AN_POPMAX               INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AF_POPMAX               FLOAT       NULLABLE   (this field was excluded)\n",
            "    > DP_MEDIAN               INTEGER     NULLABLE   (this field was excluded)\n",
            "    > DREF_MEDIAN             FLOAT       NULLABLE   (this field was excluded)\n",
            "    > GQ_MEDIAN               INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AB_MEDIAN               FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AS_RF                   FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AS_FilterStatus         STRING      NULLABLE   (4, 0, 17009588, 'PASS', 13450098, 0.7907362600434531, 'AC0', 'RF|AC0', 1, 2)\n",
            "    > AC_AFR_Male             INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AF_AMR_Male             FLOAT       NULLABLE   (this field was excluded)\n",
            "    > Hemi_NFE                INTEGER     NULLABLE   (this field was excluded)\n",
            "    > Hemi_AFR                INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AC_ASJ_Female           INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AF_FIN_Female           FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AC_OTH_Female           INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AC_NFE_Female           INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AC_EAS_Male             INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AC_OTH_Male             INTEGER     NULLABLE   (this field was excluded)\n",
            "    > Hemi_AMR                INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AC_NFE_Male             INTEGER     NULLABLE   (this field was excluded)\n",
            "    > Hemi                    INTEGER     NULLABLE   (this field was excluded)\n",
            "    > Hemi_SAS                INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AF_FIN_Male             FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AF_AFR_Male             FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AF_SAS_Male             FLOAT       NULLABLE   (this field was excluded)\n",
            "    > Hemi_ASJ                INTEGER     NULLABLE   (this field was excluded)\n",
            "    > Hemi_raw                INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AF_OTH_Male             FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AC_SAS_Female           INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AF_NFE_Female           FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AF_EAS_Female           FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AF_EAS_Male             FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AF_SAS_Female           FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AF_AFR_Female           FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AC_FIN_Female           INTEGER     NULLABLE   (this field was excluded)\n",
            "    > Hemi_OTH                INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AC_AFR_Female           INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AF_AMR_Female           FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AC_AMR_Male             INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AC_SAS_Male             INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AF_ASJ_Male             FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AC_EAS_Female           INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AC_AMR_Female           INTEGER     NULLABLE   (this field was excluded)\n",
            "    > Hemi_FIN                INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AC_FIN_Male             INTEGER     NULLABLE   (this field was excluded)\n",
            "    > AF_ASJ_Female           FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AF_OTH_Female           FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AF_NFE_Male             FLOAT       NULLABLE   (this field was excluded)\n",
            "    > AC_ASJ_Male             INTEGER     NULLABLE   (this field was excluded)\n",
            "    > Hemi_EAS                INTEGER     NULLABLE   (this field was excluded)\n",
            "    > CSQ                     RECORD      REPEATED   [68] \n",
            "        > allele              STRING      NULLABLE   (74321, 0, 136317883, 'A', 39087029, 0.2867344191370695, '-', 'TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTG', 2, 4)\n",
            "        > Consequence         STRING      NULLABLE   (231, 0, 136317883, 'downstream_gene_variant', 26360397, 0.1933744599012002, '3_prime_UTR_variant', 'upstream_gene_variant', 3, 10)\n",
            "        > IMPACT              STRING      NULLABLE   (4, 0, 136317883, 'MODIFIER', 102675373, 0.7532054543423331, 'HIGH', 'MODIFIER', 1, 3)\n",
            "        > SYMBOL              STRING      NULLABLE   (34644, 0, 136317883, '', 4843122, 0.0355281485702063, '', 'snoZ6', 2952, 12446)\n",
            "        > Gene                STRING      NULLABLE   (35494, 0, 136317883, '', 4843122, 0.0355281485702063, '', 'ENSG00000273493', 2964, 12538)\n",
            "        > Feature_type        STRING      NULLABLE   (4, 0, 136317883, 'Transcript', 131474761, 0.9644718514297937, '', 'Transcript', 1, 1)\n",
            "        > Feature             STRING      NULLABLE   (213155, 0, 136317883, 'ENST00000342175', 38434, 0.0002819439324772965, '', 'PB0182.1', 38931, 121581)\n",
            "        > BIOTYPE             STRING      NULLABLE   (43, 0, 136317883, 'protein_coding', 76696163, 0.5626273040053006, '', 'unprocessed_pseudogene', 1, 4)\n",
            "        > EXON                STRING      NULLABLE   (8497, 0, 136317883, '', 91373018, 0.670293698736504, '', '99/99', 1, 118)\n",
            "        > INTRON              STRING      NULLABLE   (8409, 0, 136317883, '', 95116329, 0.6977538596311681, '', '99/362', 1, 90)\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "4y-Ka5I8BPIE",
        "colab_type": "text"
      },
      "source": [
        "Note that although the schema identified 84 STRING fields, the output generated by exploreFieldContents above (where we skipped over all INTEGER and FLOAT fields), includes only 18 STRING fields, which means that the majority of the STRING fields are never used.\n",
        "\n",
        "Now let's see how often the **repeated** fields are repeated, by using the `exploreRepeatedFields` function defined earlier."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "jQXSaJH3Aefd",
        "colab_type": "code",
        "outputId": "65786f62-e893-4881-8d63-90d893d3babc",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 683
        }
      },
      "source": [
        "xr = bbq.bbqExploreRepeatedFields ( bqclient, \n",
        "                                    projectName, datasetName, tableName )"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "alternate_bases               RECORD      REPEATED   (6, 0, 15014473, 1, 13237344, 0.8816389359786387, 1, 6, 1, 2)\n",
            "    > CSQ                     RECORD      REPEATED   (84, 0, 17009588, 4, 1740924, 0.10234956896075319, 0, 84, 6, 16)\n",
            "names                         STRING      REPEATED   (2, 0, 15014473, 1, 9448572, 0.629297611711047, 0, 1, 1, 2)\n",
            "filter                        STRING      REPEATED   (3, 0, 15014473, 1, 13902880, 0.9259652336781984, 1, 3, 1, 1)\n",
            "call                          RECORD      REPEATED   always repeated 0 time(s)\n",
            "    > genotype                INTEGER     REPEATED   (0, 0)\n",
            "    > AD                      INTEGER     REPEATED   (0, 0)\n",
            "    > PL                      INTEGER     REPEATED   (0, 0)\n",
            "GC_AFR                        INTEGER     REPEATED   (7, 0, 15014473, 3, 12900298, 0.8591908620435762, 0, 28, 1, 2)\n",
            "GC_AMR                        INTEGER     REPEATED   (7, 0, 15014473, 3, 12900298, 0.8591908620435762, 0, 28, 1, 2)\n",
            "GC_ASJ                        INTEGER     REPEATED   (7, 0, 15014473, 3, 12900298, 0.8591908620435762, 0, 28, 1, 2)\n",
            "GC_EAS                        INTEGER     REPEATED   (7, 0, 15014473, 3, 12900298, 0.8591908620435762, 0, 28, 1, 2)\n",
            "GC_FIN                        INTEGER     REPEATED   (7, 0, 15014473, 3, 12900298, 0.8591908620435762, 0, 28, 1, 2)\n",
            "GC_NFE                        INTEGER     REPEATED   (7, 0, 15014473, 3, 12900298, 0.8591908620435762, 0, 28, 1, 2)\n",
            "GC_OTH                        INTEGER     REPEATED   (7, 0, 15014473, 3, 12900298, 0.8591908620435762, 0, 28, 1, 2)\n",
            "GC_SAS                        INTEGER     REPEATED   (7, 0, 15014473, 3, 12900298, 0.8591908620435762, 0, 28, 1, 2)\n",
            "GC_Male                       INTEGER     REPEATED   (7, 0, 15014473, 3, 13231186, 0.8812287983734094, 0, 28, 1, 2)\n",
            "GC_Female                     INTEGER     REPEATED   (7, 0, 15014473, 3, 13231186, 0.8812287983734094, 0, 28, 1, 2)\n",
            "GC_raw                        INTEGER     REPEATED   (7, 0, 15014473, 3, 13231186, 0.8812287983734094, 0, 28, 1, 2)\n",
            "GC                            INTEGER     REPEATED   (7, 0, 15014473, 3, 13231186, 0.8812287983734094, 0, 28, 1, 2)\n",
            "AS_RF_POSITIVE_TRAIN          INTEGER     REPEATED   (6, 0, 15014473, 0, 14848059, 0.9889164275029833, 0, 5, 1, 1)\n",
            "AS_RF_NEGATIVE_TRAIN          INTEGER     REPEATED   (7, 0, 15014473, 0, 14849625, 0.9890207268680026, 0, 6, 1, 1)\n",
            "GC_OTH_Male                   INTEGER     REPEATED   (7, 0, 15014473, 0, 14653035, 0.9759273602210348, 0, 28, 1, 1)\n",
            "GC_FIN_Male                   INTEGER     REPEATED   (7, 0, 15014473, 0, 14653035, 0.9759273602210348, 0, 28, 1, 1)\n",
            "GC_SAS_Male                   INTEGER     REPEATED   (7, 0, 15014473, 0, 14653035, 0.9759273602210348, 0, 28, 1, 1)\n",
            "GC_EAS_Male                   INTEGER     REPEATED   (7, 0, 15014473, 0, 14653035, 0.9759273602210348, 0, 28, 1, 1)\n",
            "GC_ASJ_Female                 INTEGER     REPEATED   (7, 0, 15014473, 0, 14653035, 0.9759273602210348, 0, 28, 1, 1)\n",
            "GC_SAS_Female                 INTEGER     REPEATED   (7, 0, 15014473, 0, 14653035, 0.9759273602210348, 0, 28, 1, 1)\n",
            "GC_ASJ_Male                   INTEGER     REPEATED   (7, 0, 15014473, 0, 14653035, 0.9759273602210348, 0, 28, 1, 1)\n",
            "GC_AFR_Female                 INTEGER     REPEATED   (7, 0, 15014473, 0, 14653035, 0.9759273602210348, 0, 28, 1, 1)\n",
            "GC_AMR_Male                   INTEGER     REPEATED   (7, 0, 15014473, 0, 14653035, 0.9759273602210348, 0, 28, 1, 1)\n",
            "GC_NFE_Male                   INTEGER     REPEATED   (7, 0, 15014473, 0, 14653035, 0.9759273602210348, 0, 28, 1, 1)\n",
            "GC_NFE_Female                 INTEGER     REPEATED   (7, 0, 15014473, 0, 14653035, 0.9759273602210348, 0, 28, 1, 1)\n",
            "GC_AMR_Female                 INTEGER     REPEATED   (7, 0, 15014473, 0, 14653035, 0.9759273602210348, 0, 28, 1, 1)\n",
            "GC_FIN_Female                 INTEGER     REPEATED   (7, 0, 15014473, 0, 14653035, 0.9759273602210348, 0, 28, 1, 1)\n",
            "GC_EAS_Female                 INTEGER     REPEATED   (7, 0, 15014473, 0, 14653035, 0.9759273602210348, 0, 28, 1, 1)\n",
            "GC_AFR_Male                   INTEGER     REPEATED   (7, 0, 15014473, 0, 14653035, 0.9759273602210348, 0, 28, 1, 1)\n",
            "GC_OTH_Female                 INTEGER     REPEATED   (7, 0, 15014473, 0, 14653035, 0.9759273602210348, 0, 28, 1, 1)\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "mkK-ewziBXvz",
        "colab_type": "text"
      },
      "source": [
        "As mentioned earlier, there are 3 records in this Variant Transform BigQuery schema:\n",
        "- **alternate_bases** is repeated anywhere between 1 and 6 times, but generally only once or twice\n",
        "- **alternate_bases.CSQ** is repeated up to 84 times, but generally 16 or fewer times\n",
        "- and the **call** record is never repeated at all, so that entire part of the schema can be ignored\n",
        "\n",
        "It might seem odd that there are no **call** records, but this is common when looking at a \"variant database\".  When looking at specific \"sample data\", on the other hand, we would expect to see details in the **call** record, including the sample name, genotype, *etc*."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "8OzK0F4vGrOZ",
        "colab_type": "text"
      },
      "source": [
        "One of the interesting fields inside the **alternate_bases.CSQ** (aka *consequences*) record is called **IMPACT** -- let's write a query to see what the different values are for that field and how often does each occur.\n",
        "\n",
        "Note that since it is a field inside a repeated record which is itself inside a repeated record, we have two implicit \"unnesting\" steps in the **FROM** clause below:\n",
        "- we first alias the long table name to `t`\n",
        "- then we implicitly unnest `t.alternate_bases` and alias it as `alt`\n",
        "- then we implicitly unnest `alt.CSQ` and alias it as `csq`\n",
        "\n",
        "That then allows us to **SELECT** `csq.IMPACT` and perform **COUNT** and **GROUP BY** operations on that field."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "1DcNxp-pGRkn",
        "colab_type": "code",
        "outputId": "f012b69c-4742-47a3-8ead-edfe41ca727c",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 173
        }
      },
      "source": [
        "%%bigquery --project isb-cgc-bq\n",
        "\n",
        "WITH\n",
        "  t1 AS (\n",
        "  SELECT\n",
        "    csq.IMPACT, COUNT(*) AS n\n",
        "  FROM\n",
        "    `bigquery-public-data.human_variant_annotation.gnomad_gnomad_exomes_hg19_release_2_0_2` AS t,\n",
        "    t.alternate_bases AS alt,\n",
        "    alt.CSQ AS csq\n",
        "  GROUP BY 1 )\n",
        "SELECT *\n",
        "FROM t1\n",
        "ORDER BY n DESC, 1"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>IMPACT</th>\n",
              "      <th>n</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>MODIFIER</td>\n",
              "      <td>102675373</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>MODERATE</td>\n",
              "      <td>18959053</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>LOW</td>\n",
              "      <td>12570156</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>HIGH</td>\n",
              "      <td>2113301</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "     IMPACT          n\n",
              "0  MODIFIER  102675373\n",
              "1  MODERATE   18959053\n",
              "2       LOW   12570156\n",
              "3      HIGH    2113301"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 10
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ivr0DE4AMr2u",
        "colab_type": "text"
      },
      "source": [
        "Ok, now let's see how many different HIGH IMPACT variants exist in this database!"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "eb1PoekmMrj_",
        "colab_type": "code",
        "outputId": "93b3cd88-25da-4bbb-fa3d-cffba98e99a3",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1989
        }
      },
      "source": [
        "%%bigquery --project isb-cgc-bq\n",
        "\n",
        "WITH\n",
        "  t1 AS (\n",
        "  SELECT DISTINCT \n",
        "    reference_name, start_position, end_position, reference_bases, AN,\n",
        "    alt.alt, alt.AC, alt.AF, alt.AF_Male, alt.AF_Female,\n",
        "    csq.allele, csq.Consequence, csq.IMPACT, csq.SYMBOL\n",
        "  FROM\n",
        "    `bigquery-public-data.human_variant_annotation.gnomad_gnomad_exomes_hg19_release_2_0_2` AS t,\n",
        "    t.alternate_bases AS alt,\n",
        "    alt.CSQ AS csq\n",
        "  WHERE\n",
        "    csq.IMPACT=\"HIGH\" ),\n",
        "  t2 AS (\n",
        "  SELECT DISTINCT \n",
        "    SYMBOL, reference_name, start_position,\n",
        "    AN, AC, AF, AF_Male, AF_Female,\n",
        "    Consequence,\n",
        "    reference_bases, alt, allele\n",
        "  FROM\n",
        "    t1\n",
        "  WHERE\n",
        "    (( AF >= 0.00001 AND AF <= 0.05000 )\n",
        "      OR ( AF >= 0.95000 AND AF <= 0.99999 ))\n",
        "    AND AN >= 10000 AND AC >= 10 )\n",
        "SELECT * FROM t2"
      ],
      "execution_count": 0,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>SYMBOL</th>\n",
              "      <th>reference_name</th>\n",
              "      <th>start_position</th>\n",
              "      <th>AN</th>\n",
              "      <th>AC</th>\n",
              "      <th>AF</th>\n",
              "      <th>AF_Male</th>\n",
              "      <th>AF_Female</th>\n",
              "      <th>Consequence</th>\n",
              "      <th>reference_bases</th>\n",
              "      <th>alt</th>\n",
              "      <th>allele</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>GOLGA6L6</td>\n",
              "      <td>15</td>\n",
              "      <td>20740192</td>\n",
              "      <td>137056</td>\n",
              "      <td>144</td>\n",
              "      <td>0.001051</td>\n",
              "      <td>0.000988</td>\n",
              "      <td>0.001124</td>\n",
              "      <td>frameshift_variant</td>\n",
              "      <td>T</td>\n",
              "      <td>TGC</td>\n",
              "      <td>GC</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>CDCP2</td>\n",
              "      <td>1</td>\n",
              "      <td>54605317</td>\n",
              "      <td>152972</td>\n",
              "      <td>789</td>\n",
              "      <td>0.005158</td>\n",
              "      <td>0.005591</td>\n",
              "      <td>0.004625</td>\n",
              "      <td>frameshift_variant</td>\n",
              "      <td>TG</td>\n",
              "      <td>TGCG</td>\n",
              "      <td>CG</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>FAM83C</td>\n",
              "      <td>20</td>\n",
              "      <td>33874804</td>\n",
              "      <td>242996</td>\n",
              "      <td>16</td>\n",
              "      <td>0.000066</td>\n",
              "      <td>0.000067</td>\n",
              "      <td>0.000064</td>\n",
              "      <td>frameshift_variant</td>\n",
              "      <td>G</td>\n",
              "      <td>GGCCAC</td>\n",
              "      <td>GCCAC</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>CD36</td>\n",
              "      <td>7</td>\n",
              "      <td>80302137</td>\n",
              "      <td>244370</td>\n",
              "      <td>14</td>\n",
              "      <td>0.000057</td>\n",
              "      <td>0.000067</td>\n",
              "      <td>0.000045</td>\n",
              "      <td>frameshift_variant</td>\n",
              "      <td>T</td>\n",
              "      <td>TCAAGC</td>\n",
              "      <td>CAAGC</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>AC064874.1</td>\n",
              "      <td>2</td>\n",
              "      <td>236682978</td>\n",
              "      <td>143494</td>\n",
              "      <td>70</td>\n",
              "      <td>0.000488</td>\n",
              "      <td>0.000615</td>\n",
              "      <td>0.000336</td>\n",
              "      <td>frameshift_variant</td>\n",
              "      <td>T</td>\n",
              "      <td>TGCTG</td>\n",
              "      <td>GCTG</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>KRTAP2-4</td>\n",
              "      <td>17</td>\n",
              "      <td>39221826</td>\n",
              "      <td>127396</td>\n",
              "      <td>26</td>\n",
              "      <td>0.000204</td>\n",
              "      <td>0.000115</td>\n",
              "      <td>0.000311</td>\n",
              "      <td>frameshift_variant</td>\n",
              "      <td>GCC</td>\n",
              "      <td>GGTCC</td>\n",
              "      <td>GT</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>RBM12B</td>\n",
              "      <td>8</td>\n",
              "      <td>94746697</td>\n",
              "      <td>245754</td>\n",
              "      <td>24</td>\n",
              "      <td>0.000098</td>\n",
              "      <td>0.000096</td>\n",
              "      <td>0.000099</td>\n",
              "      <td>frameshift_variant</td>\n",
              "      <td>G</td>\n",
              "      <td>GGGAGCTGCCTGAAGTCCTCCTCT</td>\n",
              "      <td>GGAGCTGCCTGAAGTCCTCCTCT</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>TCHH</td>\n",
              "      <td>1</td>\n",
              "      <td>152084455</td>\n",
              "      <td>123622</td>\n",
              "      <td>47</td>\n",
              "      <td>0.000380</td>\n",
              "      <td>0.000219</td>\n",
              "      <td>0.000581</td>\n",
              "      <td>frameshift_variant</td>\n",
              "      <td>G</td>\n",
              "      <td>GTCCTC</td>\n",
              "      <td>TCCTC</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>CMYA5</td>\n",
              "      <td>5</td>\n",
              "      <td>79025349</td>\n",
              "      <td>244398</td>\n",
              "      <td>11</td>\n",
              "      <td>0.000045</td>\n",
              "      <td>0.000082</td>\n",
              "      <td>0.000000</td>\n",
              "      <td>frameshift_variant</td>\n",
              "      <td>T</td>\n",
              "      <td>TAA</td>\n",
              "      <td>AA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>HAGHL</td>\n",
              "      <td>16</td>\n",
              "      <td>778960</td>\n",
              "      <td>163440</td>\n",
              "      <td>81</td>\n",
              "      <td>0.000496</td>\n",
              "      <td>0.000424</td>\n",
              "      <td>0.000582</td>\n",
              "      <td>frameshift_variant</td>\n",
              "      <td>G</td>\n",
              "      <td>GGC</td>\n",
              "      <td>GC</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>10</th>\n",
              "      <td>MRGPRF</td>\n",
              "      <td>11</td>\n",
              "      <td>68777129</td>\n",
              "      <td>122702</td>\n",
              "      <td>292</td>\n",
              "      <td>0.002380</td>\n",
              "      <td>0.002071</td>\n",
              "      <td>0.002753</td>\n",
              "      <td>frameshift_variant</td>\n",
              "      <td>C</td>\n",
              "      <td>CAGGAGGGT</td>\n",
              "      <td>AGGAGGGT</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11</th>\n",
              "      <td>ZNF880</td>\n",
              "      <td>19</td>\n",
              "      <td>52888069</td>\n",
              "      <td>206844</td>\n",
              "      <td>297</td>\n",
              "      <td>0.001436</td>\n",
              "      <td>0.001229</td>\n",
              "      <td>0.001686</td>\n",
              "      <td>stop_gained&amp;frameshift_variant</td>\n",
              "      <td>T</td>\n",
              "      <td>TTTGGGAGGCCGAGGCGGGTGGATCATGAGGTCAGGAGATCGAGAC...</td>\n",
              "      <td>TTGGGAGGCCGAGGCGGGTGGATCATGAGGTCAGGAGATCGAGACC...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>12</th>\n",
              "      <td>VWDE</td>\n",
              "      <td>7</td>\n",
              "      <td>12428956</td>\n",
              "      <td>151298</td>\n",
              "      <td>13</td>\n",
              "      <td>0.000086</td>\n",
              "      <td>0.000099</td>\n",
              "      <td>0.000071</td>\n",
              "      <td>frameshift_variant&amp;NMD_transcript_variant</td>\n",
              "      <td>T</td>\n",
              "      <td>TGG</td>\n",
              "      <td>GG</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>13</th>\n",
              "      <td>MUC2</td>\n",
              "      <td>11</td>\n",
              "      <td>1092807</td>\n",
              "      <td>141764</td>\n",
              "      <td>12</td>\n",
              "      <td>0.000085</td>\n",
              "      <td>0.000092</td>\n",
              "      <td>0.000076</td>\n",
              "      <td>frameshift_variant</td>\n",
              "      <td>A</td>\n",
              "      <td>AATGACTACGC</td>\n",
              "      <td>ATGACTACGC</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>14</th>\n",
              "      <td>PRDM15</td>\n",
              "      <td>21</td>\n",
              "      <td>43298844</td>\n",
              "      <td>185760</td>\n",
              "      <td>26</td>\n",
              "      <td>0.000140</td>\n",
              "      <td>0.000085</td>\n",
              "      <td>0.000212</td>\n",
              "      <td>frameshift_variant&amp;NMD_transcript_variant</td>\n",
              "      <td>AC</td>\n",
              "      <td>ACCC</td>\n",
              "      <td>CC</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>HLA-DRB1</td>\n",
              "      <td>6</td>\n",
              "      <td>32551959</td>\n",
              "      <td>137856</td>\n",
              "      <td>2873</td>\n",
              "      <td>0.020841</td>\n",
              "      <td>0.023721</td>\n",
              "      <td>0.017225</td>\n",
              "      <td>frameshift_variant</td>\n",
              "      <td>TGC</td>\n",
              "      <td>TCCGC</td>\n",
              "      <td>CC</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>16</th>\n",
              "      <td>NFKBIZ</td>\n",
              "      <td>3</td>\n",
              "      <td>101578208</td>\n",
              "      <td>246014</td>\n",
              "      <td>32</td>\n",
              "      <td>0.000130</td>\n",
              "      <td>0.000134</td>\n",
              "      <td>0.000126</td>\n",
              "      <td>frameshift_variant</td>\n",
              "      <td>G</td>\n",
              "      <td>GTA</td>\n",
              "      <td>TA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>KRR1</td>\n",
              "      <td>12</td>\n",
              "      <td>75900694</td>\n",
              "      <td>219556</td>\n",
              "      <td>127</td>\n",
              "      <td>0.000578</td>\n",
              "      <td>0.000521</td>\n",
              "      <td>0.000648</td>\n",
              "      <td>stop_gained&amp;frameshift_variant&amp;splice_region_v...</td>\n",
              "      <td>T</td>\n",
              "      <td>TGATGTTCATTTAAG</td>\n",
              "      <td>GATGTTCATTTAAG</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>18</th>\n",
              "      <td>CSAD</td>\n",
              "      <td>12</td>\n",
              "      <td>53566206</td>\n",
              "      <td>246256</td>\n",
              "      <td>31</td>\n",
              "      <td>0.000126</td>\n",
              "      <td>0.000082</td>\n",
              "      <td>0.000180</td>\n",
              "      <td>frameshift_variant</td>\n",
              "      <td>G</td>\n",
              "      <td>GAACC</td>\n",
              "      <td>AACC</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>19</th>\n",
              "      <td>THEG</td>\n",
              "      <td>19</td>\n",
              "      <td>374527</td>\n",
              "      <td>114814</td>\n",
              "      <td>12</td>\n",
              "      <td>0.000105</td>\n",
              "      <td>0.000132</td>\n",
              "      <td>0.000074</td>\n",
              "      <td>frameshift_variant</td>\n",
              "      <td>C</td>\n",
              "      <td>CCT</td>\n",
              "      <td>CT</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>20</th>\n",
              "      <td>BOD1</td>\n",
              "      <td>5</td>\n",
              "      <td>173036437</td>\n",
              "      <td>198302</td>\n",
              "      <td>51</td>\n",
              "      <td>0.000257</td>\n",
              "      <td>0.000312</td>\n",
              "      <td>0.000190</td>\n",
              "      <td>splice_acceptor_variant</td>\n",
              "      <td>C</td>\n",
              "      <td>CCACACTCTGCCTCAAACCATTTCGCAA</td>\n",
              "      <td>CACACTCTGCCTCAAACCATTTCGCAA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>21</th>\n",
              "      <td>ACP2</td>\n",
              "      <td>11</td>\n",
              "      <td>47266418</td>\n",
              "      <td>236294</td>\n",
              "      <td>32</td>\n",
              "      <td>0.000135</td>\n",
              "      <td>0.000115</td>\n",
              "      <td>0.000160</td>\n",
              "      <td>splice_acceptor_variant&amp;non_coding_transcript_...</td>\n",
              "      <td>C</td>\n",
              "      <td>CTG</td>\n",
              "      <td>TG</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>22</th>\n",
              "      <td>PDLIM1</td>\n",
              "      <td>10</td>\n",
              "      <td>96997866</td>\n",
              "      <td>242686</td>\n",
              "      <td>24</td>\n",
              "      <td>0.000099</td>\n",
              "      <td>0.000098</td>\n",
              "      <td>0.000100</td>\n",
              "      <td>frameshift_variant&amp;splice_region_variant</td>\n",
              "      <td>C</td>\n",
              "      <td>CAACAATCCCAGTGCCACATTTGTCACACATAGGCA</td>\n",
              "      <td>AACAATCCCAGTGCCACATTTGTCACACATAGGCA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>23</th>\n",
              "      <td>APOL3</td>\n",
              "      <td>22</td>\n",
              "      <td>36556872</td>\n",
              "      <td>246006</td>\n",
              "      <td>19</td>\n",
              "      <td>0.000077</td>\n",
              "      <td>0.000089</td>\n",
              "      <td>0.000063</td>\n",
              "      <td>frameshift_variant&amp;NMD_transcript_variant</td>\n",
              "      <td>G</td>\n",
              "      <td>GATTCACC</td>\n",
              "      <td>ATTCACC</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24</th>\n",
              "      <td>PDLIM1</td>\n",
              "      <td>10</td>\n",
              "      <td>96997867</td>\n",
              "      <td>242536</td>\n",
              "      <td>17</td>\n",
              "      <td>0.000070</td>\n",
              "      <td>0.000083</td>\n",
              "      <td>0.000055</td>\n",
              "      <td>frameshift_variant&amp;splice_region_variant</td>\n",
              "      <td>A</td>\n",
              "      <td>AACAATCCCAGTGCCACATTTGTCACACATAGGCAACTTCTGAGCA...</td>\n",
              "      <td>ACAATCCCAGTGCCACATTTGTCACACATAGGCAACTTCTGAGCAT...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>25</th>\n",
              "      <td>LPAR6</td>\n",
              "      <td>13</td>\n",
              "      <td>48986197</td>\n",
              "      <td>219506</td>\n",
              "      <td>14</td>\n",
              "      <td>0.000064</td>\n",
              "      <td>0.000058</td>\n",
              "      <td>0.000071</td>\n",
              "      <td>frameshift_variant</td>\n",
              "      <td>G</td>\n",
              "      <td>GATAC</td>\n",
              "      <td>ATAC</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>26</th>\n",
              "      <td>MUC4</td>\n",
              "      <td>3</td>\n",
              "      <td>195507035</td>\n",
              "      <td>94934</td>\n",
              "      <td>572</td>\n",
              "      <td>0.006025</td>\n",
              "      <td>0.006447</td>\n",
              "      <td>0.005518</td>\n",
              "      <td>frameshift_variant&amp;NMD_transcript_variant</td>\n",
              "      <td>C</td>\n",
              "      <td>CCTGTGGATGCTGCGGAAGTGTTGGTGACATGAAGAGGGGTGGCAT...</td>\n",
              "      <td>CTGTGGATGCTGCGGAAGTGTTGGTGACATGAAGAGGGGTGGCATG...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>27</th>\n",
              "      <td>C16orf3</td>\n",
              "      <td>16</td>\n",
              "      <td>90095595</td>\n",
              "      <td>160372</td>\n",
              "      <td>56</td>\n",
              "      <td>0.000349</td>\n",
              "      <td>0.000434</td>\n",
              "      <td>0.000247</td>\n",
              "      <td>frameshift_variant</td>\n",
              "      <td>A</td>\n",
              "      <td>ACGGGGCAGCC</td>\n",
              "      <td>CGGGGCAGCC</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>28</th>\n",
              "      <td>KIAA1468</td>\n",
              "      <td>18</td>\n",
              "      <td>59942745</td>\n",
              "      <td>108430</td>\n",
              "      <td>1254</td>\n",
              "      <td>0.011565</td>\n",
              "      <td>0.011634</td>\n",
              "      <td>0.011480</td>\n",
              "      <td>frameshift_variant</td>\n",
              "      <td>ATTT</td>\n",
              "      <td>ATTTTT</td>\n",
              "      <td>TT</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>29</th>\n",
              "      <td>MUC4</td>\n",
              "      <td>3</td>\n",
              "      <td>195510345</td>\n",
              "      <td>132306</td>\n",
              "      <td>22</td>\n",
              "      <td>0.000166</td>\n",
              "      <td>0.000099</td>\n",
              "      <td>0.000243</td>\n",
              "      <td>frameshift_variant&amp;NMD_transcript_variant</td>\n",
              "      <td>GTGTCA</td>\n",
              "      <td>GCATGTCA</td>\n",
              "      <td>CA</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53588</th>\n",
              "      <td>NUFIP1</td>\n",
              "      <td>13</td>\n",
              "      <td>45556206</td>\n",
              "      <td>183792</td>\n",
              "      <td>17</td>\n",
              "      <td>0.000092</td>\n",
              "      <td>0.000127</td>\n",
              "      <td>0.000049</td>\n",
              "      <td>frameshift_variant&amp;splice_region_variant</td>\n",
              "      <td>CT</td>\n",
              "      <td>CTT</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53589</th>\n",
              "      <td>CEP290</td>\n",
              "      <td>12</td>\n",
              "      <td>88524077</td>\n",
              "      <td>232776</td>\n",
              "      <td>22</td>\n",
              "      <td>0.000095</td>\n",
              "      <td>0.000102</td>\n",
              "      <td>0.000086</td>\n",
              "      <td>frameshift_variant&amp;NMD_transcript_variant</td>\n",
              "      <td>GT</td>\n",
              "      <td>GTT</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53590</th>\n",
              "      <td>CHIA</td>\n",
              "      <td>1</td>\n",
              "      <td>111857218</td>\n",
              "      <td>246036</td>\n",
              "      <td>35</td>\n",
              "      <td>0.000142</td>\n",
              "      <td>0.000119</td>\n",
              "      <td>0.000171</td>\n",
              "      <td>splice_donor_variant&amp;NMD_transcript_variant</td>\n",
              "      <td>G</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53591</th>\n",
              "      <td>ADH6</td>\n",
              "      <td>4</td>\n",
              "      <td>100130022</td>\n",
              "      <td>245582</td>\n",
              "      <td>42</td>\n",
              "      <td>0.000171</td>\n",
              "      <td>0.000216</td>\n",
              "      <td>0.000117</td>\n",
              "      <td>start_lost</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53592</th>\n",
              "      <td>AC079354.1</td>\n",
              "      <td>2</td>\n",
              "      <td>202967837</td>\n",
              "      <td>143484</td>\n",
              "      <td>41</td>\n",
              "      <td>0.000286</td>\n",
              "      <td>0.000308</td>\n",
              "      <td>0.000259</td>\n",
              "      <td>splice_donor_variant</td>\n",
              "      <td>G</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53593</th>\n",
              "      <td>MYO3B</td>\n",
              "      <td>2</td>\n",
              "      <td>171225756</td>\n",
              "      <td>245574</td>\n",
              "      <td>10</td>\n",
              "      <td>0.000041</td>\n",
              "      <td>0.000059</td>\n",
              "      <td>0.000018</td>\n",
              "      <td>stop_gained&amp;NMD_transcript_variant</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53594</th>\n",
              "      <td>BPIFA3</td>\n",
              "      <td>20</td>\n",
              "      <td>31811615</td>\n",
              "      <td>245444</td>\n",
              "      <td>21</td>\n",
              "      <td>0.000086</td>\n",
              "      <td>0.000060</td>\n",
              "      <td>0.000117</td>\n",
              "      <td>splice_acceptor_variant&amp;non_coding_transcript_...</td>\n",
              "      <td>G</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53595</th>\n",
              "      <td>TSEN54</td>\n",
              "      <td>17</td>\n",
              "      <td>73517589</td>\n",
              "      <td>246076</td>\n",
              "      <td>20</td>\n",
              "      <td>0.000081</td>\n",
              "      <td>0.000052</td>\n",
              "      <td>0.000117</td>\n",
              "      <td>stop_gained&amp;splice_region_variant&amp;NMD_transcri...</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53596</th>\n",
              "      <td>ZNF544</td>\n",
              "      <td>19</td>\n",
              "      <td>58773814</td>\n",
              "      <td>246234</td>\n",
              "      <td>4180</td>\n",
              "      <td>0.016976</td>\n",
              "      <td>0.017385</td>\n",
              "      <td>0.016480</td>\n",
              "      <td>stop_gained</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53597</th>\n",
              "      <td>MLKL</td>\n",
              "      <td>16</td>\n",
              "      <td>74725173</td>\n",
              "      <td>241160</td>\n",
              "      <td>26</td>\n",
              "      <td>0.000108</td>\n",
              "      <td>0.000106</td>\n",
              "      <td>0.000110</td>\n",
              "      <td>splice_donor_variant&amp;NMD_transcript_variant</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53598</th>\n",
              "      <td>TTC23L</td>\n",
              "      <td>5</td>\n",
              "      <td>34850370</td>\n",
              "      <td>246150</td>\n",
              "      <td>117</td>\n",
              "      <td>0.000475</td>\n",
              "      <td>0.000527</td>\n",
              "      <td>0.000413</td>\n",
              "      <td>stop_gained</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53599</th>\n",
              "      <td>SMYD3</td>\n",
              "      <td>1</td>\n",
              "      <td>246518331</td>\n",
              "      <td>246176</td>\n",
              "      <td>11</td>\n",
              "      <td>0.000045</td>\n",
              "      <td>0.000022</td>\n",
              "      <td>0.000072</td>\n",
              "      <td>splice_donor_variant</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53600</th>\n",
              "      <td>DRAM2</td>\n",
              "      <td>1</td>\n",
              "      <td>111661441</td>\n",
              "      <td>240852</td>\n",
              "      <td>11</td>\n",
              "      <td>0.000046</td>\n",
              "      <td>0.000038</td>\n",
              "      <td>0.000055</td>\n",
              "      <td>stop_gained&amp;frameshift_variant</td>\n",
              "      <td>G</td>\n",
              "      <td>GT</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53601</th>\n",
              "      <td>GPR125</td>\n",
              "      <td>4</td>\n",
              "      <td>22444484</td>\n",
              "      <td>243210</td>\n",
              "      <td>14</td>\n",
              "      <td>0.000058</td>\n",
              "      <td>0.000060</td>\n",
              "      <td>0.000054</td>\n",
              "      <td>stop_gained&amp;splice_region_variant</td>\n",
              "      <td>G</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53602</th>\n",
              "      <td>XIRP2</td>\n",
              "      <td>2</td>\n",
              "      <td>168107295</td>\n",
              "      <td>245346</td>\n",
              "      <td>12</td>\n",
              "      <td>0.000049</td>\n",
              "      <td>0.000052</td>\n",
              "      <td>0.000045</td>\n",
              "      <td>stop_gained</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53603</th>\n",
              "      <td>OR5B17</td>\n",
              "      <td>11</td>\n",
              "      <td>58126152</td>\n",
              "      <td>245774</td>\n",
              "      <td>36</td>\n",
              "      <td>0.000146</td>\n",
              "      <td>0.000119</td>\n",
              "      <td>0.000180</td>\n",
              "      <td>stop_gained</td>\n",
              "      <td>A</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53604</th>\n",
              "      <td>VPS13B</td>\n",
              "      <td>8</td>\n",
              "      <td>100520085</td>\n",
              "      <td>246078</td>\n",
              "      <td>10</td>\n",
              "      <td>0.000041</td>\n",
              "      <td>0.000037</td>\n",
              "      <td>0.000045</td>\n",
              "      <td>stop_gained</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53605</th>\n",
              "      <td>DNAH17</td>\n",
              "      <td>17</td>\n",
              "      <td>76476763</td>\n",
              "      <td>245304</td>\n",
              "      <td>41</td>\n",
              "      <td>0.000167</td>\n",
              "      <td>0.000164</td>\n",
              "      <td>0.000171</td>\n",
              "      <td>splice_donor_variant</td>\n",
              "      <td>A</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53606</th>\n",
              "      <td>GPR148</td>\n",
              "      <td>2</td>\n",
              "      <td>131486964</td>\n",
              "      <td>246100</td>\n",
              "      <td>60</td>\n",
              "      <td>0.000244</td>\n",
              "      <td>0.000200</td>\n",
              "      <td>0.000297</td>\n",
              "      <td>stop_gained</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53607</th>\n",
              "      <td>WWOX</td>\n",
              "      <td>16</td>\n",
              "      <td>79245695</td>\n",
              "      <td>245286</td>\n",
              "      <td>1858</td>\n",
              "      <td>0.007575</td>\n",
              "      <td>0.005684</td>\n",
              "      <td>0.009875</td>\n",
              "      <td>stop_gained</td>\n",
              "      <td>G</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53608</th>\n",
              "      <td>GOLGA6L1</td>\n",
              "      <td>15</td>\n",
              "      <td>22743356</td>\n",
              "      <td>125624</td>\n",
              "      <td>36</td>\n",
              "      <td>0.000287</td>\n",
              "      <td>0.000308</td>\n",
              "      <td>0.000261</td>\n",
              "      <td>frameshift_variant</td>\n",
              "      <td>G</td>\n",
              "      <td>GT</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53609</th>\n",
              "      <td>STRC</td>\n",
              "      <td>15</td>\n",
              "      <td>43896350</td>\n",
              "      <td>228588</td>\n",
              "      <td>11</td>\n",
              "      <td>0.000048</td>\n",
              "      <td>0.000048</td>\n",
              "      <td>0.000049</td>\n",
              "      <td>splice_acceptor_variant</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53610</th>\n",
              "      <td>FLG</td>\n",
              "      <td>1</td>\n",
              "      <td>152281122</td>\n",
              "      <td>246264</td>\n",
              "      <td>18</td>\n",
              "      <td>0.000073</td>\n",
              "      <td>0.000067</td>\n",
              "      <td>0.000081</td>\n",
              "      <td>stop_gained</td>\n",
              "      <td>G</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53611</th>\n",
              "      <td>MEP1B</td>\n",
              "      <td>18</td>\n",
              "      <td>29797784</td>\n",
              "      <td>246128</td>\n",
              "      <td>26</td>\n",
              "      <td>0.000106</td>\n",
              "      <td>0.000082</td>\n",
              "      <td>0.000135</td>\n",
              "      <td>stop_gained</td>\n",
              "      <td>CGA</td>\n",
              "      <td>TGA</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53612</th>\n",
              "      <td>PCDHB1</td>\n",
              "      <td>5</td>\n",
              "      <td>140433284</td>\n",
              "      <td>246020</td>\n",
              "      <td>26</td>\n",
              "      <td>0.000106</td>\n",
              "      <td>0.000096</td>\n",
              "      <td>0.000117</td>\n",
              "      <td>stop_gained</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53613</th>\n",
              "      <td>SARS2</td>\n",
              "      <td>19</td>\n",
              "      <td>39440432</td>\n",
              "      <td>27352</td>\n",
              "      <td>83</td>\n",
              "      <td>0.003035</td>\n",
              "      <td>0.003226</td>\n",
              "      <td>0.002826</td>\n",
              "      <td>splice_donor_variant</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53614</th>\n",
              "      <td>RABL6</td>\n",
              "      <td>9</td>\n",
              "      <td>139735150</td>\n",
              "      <td>146334</td>\n",
              "      <td>75</td>\n",
              "      <td>0.000513</td>\n",
              "      <td>0.000469</td>\n",
              "      <td>0.000563</td>\n",
              "      <td>stop_gained</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53615</th>\n",
              "      <td>SEPT3</td>\n",
              "      <td>22</td>\n",
              "      <td>42390345</td>\n",
              "      <td>245736</td>\n",
              "      <td>29</td>\n",
              "      <td>0.000118</td>\n",
              "      <td>0.000104</td>\n",
              "      <td>0.000135</td>\n",
              "      <td>splice_donor_variant&amp;NMD_transcript_variant</td>\n",
              "      <td>G</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53616</th>\n",
              "      <td>MEGF9</td>\n",
              "      <td>9</td>\n",
              "      <td>123453686</td>\n",
              "      <td>179724</td>\n",
              "      <td>14</td>\n",
              "      <td>0.000078</td>\n",
              "      <td>0.000092</td>\n",
              "      <td>0.000061</td>\n",
              "      <td>splice_donor_variant</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>53617</th>\n",
              "      <td>EPHA8</td>\n",
              "      <td>1</td>\n",
              "      <td>22915707</td>\n",
              "      <td>243376</td>\n",
              "      <td>15</td>\n",
              "      <td>0.000062</td>\n",
              "      <td>0.000097</td>\n",
              "      <td>0.000018</td>\n",
              "      <td>stop_gained</td>\n",
              "      <td>A</td>\n",
              "      <td>T</td>\n",
              "      <td>T</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>53618 rows × 12 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "           SYMBOL  ...                                             allele\n",
              "0        GOLGA6L6  ...                                                 GC\n",
              "1           CDCP2  ...                                                 CG\n",
              "2          FAM83C  ...                                              GCCAC\n",
              "3            CD36  ...                                              CAAGC\n",
              "4      AC064874.1  ...                                               GCTG\n",
              "5        KRTAP2-4  ...                                                 GT\n",
              "6          RBM12B  ...                            GGAGCTGCCTGAAGTCCTCCTCT\n",
              "7            TCHH  ...                                              TCCTC\n",
              "8           CMYA5  ...                                                 AA\n",
              "9           HAGHL  ...                                                 GC\n",
              "10         MRGPRF  ...                                           AGGAGGGT\n",
              "11         ZNF880  ...  TTGGGAGGCCGAGGCGGGTGGATCATGAGGTCAGGAGATCGAGACC...\n",
              "12           VWDE  ...                                                 GG\n",
              "13           MUC2  ...                                         ATGACTACGC\n",
              "14         PRDM15  ...                                                 CC\n",
              "15       HLA-DRB1  ...                                                 CC\n",
              "16         NFKBIZ  ...                                                 TA\n",
              "17           KRR1  ...                                     GATGTTCATTTAAG\n",
              "18           CSAD  ...                                               AACC\n",
              "19           THEG  ...                                                 CT\n",
              "20           BOD1  ...                        CACACTCTGCCTCAAACCATTTCGCAA\n",
              "21           ACP2  ...                                                 TG\n",
              "22         PDLIM1  ...                AACAATCCCAGTGCCACATTTGTCACACATAGGCA\n",
              "23          APOL3  ...                                            ATTCACC\n",
              "24         PDLIM1  ...  ACAATCCCAGTGCCACATTTGTCACACATAGGCAACTTCTGAGCAT...\n",
              "25          LPAR6  ...                                               ATAC\n",
              "26           MUC4  ...  CTGTGGATGCTGCGGAAGTGTTGGTGACATGAAGAGGGGTGGCATG...\n",
              "27        C16orf3  ...                                         CGGGGCAGCC\n",
              "28       KIAA1468  ...                                                 TT\n",
              "29           MUC4  ...                                                 CA\n",
              "...           ...  ...                                                ...\n",
              "53588      NUFIP1  ...                                                  T\n",
              "53589      CEP290  ...                                                  T\n",
              "53590        CHIA  ...                                                  T\n",
              "53591        ADH6  ...                                                  T\n",
              "53592  AC079354.1  ...                                                  T\n",
              "53593       MYO3B  ...                                                  T\n",
              "53594      BPIFA3  ...                                                  T\n",
              "53595      TSEN54  ...                                                  T\n",
              "53596      ZNF544  ...                                                  T\n",
              "53597        MLKL  ...                                                  T\n",
              "53598      TTC23L  ...                                                  T\n",
              "53599       SMYD3  ...                                                  T\n",
              "53600       DRAM2  ...                                                  T\n",
              "53601      GPR125  ...                                                  T\n",
              "53602       XIRP2  ...                                                  T\n",
              "53603      OR5B17  ...                                                  T\n",
              "53604      VPS13B  ...                                                  T\n",
              "53605      DNAH17  ...                                                  T\n",
              "53606      GPR148  ...                                                  T\n",
              "53607        WWOX  ...                                                  T\n",
              "53608    GOLGA6L1  ...                                                  T\n",
              "53609        STRC  ...                                                  T\n",
              "53610         FLG  ...                                                  T\n",
              "53611       MEP1B  ...                                                  T\n",
              "53612      PCDHB1  ...                                                  T\n",
              "53613       SARS2  ...                                                  T\n",
              "53614       RABL6  ...                                                  T\n",
              "53615       SEPT3  ...                                                  T\n",
              "53616       MEGF9  ...                                                  T\n",
              "53617       EPHA8  ...                                                  T\n",
              "\n",
              "[53618 rows x 12 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 11
        }
      ]
    }
  ]
}